---
title: "Make lncRNA Table"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(readr)
library(dplyr)
library(matrixStats)
library(rtracklayer)
library(here)
```

```{r}
data_base_dir= "/space/rnaseq_enrichment/workspace/lncpipe/rnaseh_tot_rz/lncpipe_out/Result"
input_gtf= file.path(data_base_dir, "Identified_lncRNA/final_all.gtf")
tpm_tsv= file.path(data_base_dir,"Quantification/kallisto.tpm.txt")
out_table_csv = "/space/rnaseq_enrichment/workspace/lncpipe/lncrna_summary_table.csv"
novel_lncrna_gtf = here("info","novel_lncrna.gtf")

tpm_tsv %>%
  read_tsv %>%
  filter(Type=="novel") ->
  novel_tpm_df
```

# Summarize GTF Data
```{r}
import(input_gtf, format="gtf") %>%
  subset(source=="novel") ->
  novel_lncrna_granges
novel_lncrna_granges
```

```{r}
novel_lncrna_granges %>%
  as.data.frame %>%
  group_by(gene_id) %>%
  summarise(num_exons = n(), gene_start=min(start), gene_end=max(end), chrom=seqnames[1], tot_exon_len=sum(width)) ->
  gtf_summary_df

gtf_summary_df

# This runs fine in RStudio, but knitting dies in this chunk (in group_by?) with "check_dots_used”
```

## Export GTF with novel lncRNAs
```{r}
export(novel_lncrna_granges, novel_lncrna_gtf,format="gtf")
```

# Summarize Kallisto TPM Data
```{r}
novel_tpm_df %>%
  mutate(tpm_sum=rowSums(select_at(., vars(starts_with("X")))),
         tpm_mean=rowMeans(select_at(., vars(starts_with("X")))),
         tpm_median=rowMedians(as.matrix(select_at(., vars(starts_with("X")))))
         ) %>%
  dplyr::select(ID, Type, tpm_mean, tpm_median) ->
  tpm_summary_df

tpm_summary_df
```


# Combine Tables
```{r}
full_join(tpm_summary_df, gtf_summary_df, by=c("ID"="gene_id")) %>%
  arrange(chrom, gene_start) %>%
  dplyr::select(Name=ID,
                Chromosome=chrom,
                Start=gene_start,
                End=gene_end,
                `# Exons`=num_exons,
                gene_start,gene_end,chrom,
                `Total Exonic Length`=tot_exon_len,
                `Mean TPM`=tpm_mean,
                `Median TPM`=tpm_median
                ) ->
  final_lncrna_table

final_lncrna_table %>%
  write_csv(out_table_csv)

final_lncrna_table

# "ID,Type,tpm_mean,tpm_median,num_exons,gene_start,gene_end,chrom,tot_exon_len"
```

## Mean/Median TPM across all samples?
## Rank among all ncRNA (for Average TPM)?
## Rank among all genes (for Average TPM)?
## Total length of the lncrna
## number of exons
## chrom, start, stop


