---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r}
library(dplyr)
library(fs)
library(R.utils)
library(here)
library(readr)
library(stringr)
```

# Setup
## Make functions
```{r}
ensembl_download <- function(url, download_dir) {
  url %>%
    basename %>%
    file.path(download_dir, .) ->
    gz_filename
  
  gz_filename %>%
    path_ext_remove ->
    unziped_filename
  
  dir_create(download_dir, recurse = TRUE)
  download.file(url, gz_filename)
  gunzip(gz_filename)
  warning("\n-----should test ensembl checksums------\n")

  return(unziped_filename)
  }
```

## Paths
```{r}
source("../common_config.R")

basedir = "/space"
scratch_dir=file.path(basedir, "lncpipe/rnaseh_only_genome_39")
bash_script=file.path(scratch_dir,"lncpipe_prep.sh")
genome_dir=file.path(scratch_dir, "genome")
lncpipe_dir=file.path(scratch_dir, "lncpipe")
lncpipe_out_dir=file.path(scratch_dir, "lncpipe_out")
if(dir_exists(scratch_dir)) {
  dir_delete(scratch_dir)
}
dir_create(lncpipe_out_dir)

hexamer_path = file.path(genome_dir,"h99_hexamer.tsv")
logit_model_prefix = file.path(genome_dir,"h99")
logit_model_path = paste0(logit_model_prefix, ".logit.RData")

fastq_basedir = file.path(basedir, "hts2018/rawdata")
fastq_2018_dir = file.path(fastq_basedir, "hts_2018_data/hts2018_pilot_rawdata")
fastq_2019_dir = file.path(fastq_basedir, "hts_2019_data/hts2019_pilot_rawdata")
fastq_link_dir = file.path(scratch_dir, "fastq_links")
dir_create(fastq_link_dir)
```

## Set up links to FASTQs
need to allow complex selection of fastqs
### Run on RNaseH, Ribo-Zero, and Total
```{r eval=FALSE, include=FALSE}
# get the 2018 files, except the polyA enriched (MA)
list.files(fastq_2018_dir, pattern = ".fastq.gz", full.names = TRUE) %>%
  str_subset("_MA_", negate = TRUE) ->
  fastq_2018_paths

list.files(fastq_2019_dir, pattern = ".fastq.gz", full.names = TRUE) %>%
  str_subset("_M1_", negate = TRUE) ->
  fastq_2019_paths

orig_fastqs = c(fastq_2018_paths, fastq_2019_paths)

orig_fastqs %>%
  basename %>%
  file.path(fastq_link_dir,.) ->
  link_paths

link_create(orig_fastqs, link_paths, symbolic = FALSE)
```

## Run on RNaseH only
```{r}
list.files(fastq_2019_dir, pattern = "_2018_P_H", full.names = TRUE) ->
  fastq_2019_paths

orig_fastqs = c(fastq_2019_paths)

orig_fastqs %>%
  basename %>%
  file.path(fastq_link_dir,.) ->
  link_paths

link_create(orig_fastqs, link_paths, symbolic = FALSE)
```

## Download data
```{r}
ncrna_path = ensembl_download(h99_ncrna_url, genome_dir)
cds_path = ensembl_download(h99_cds_url, genome_dir)
genome_path = ensembl_download(h99_genome_url, genome_dir)
orig_gtf_path = ensembl_download(h99_gtf_url, genome_dir)

# orig_gtf_path %>%
#   path_ext_set("hacked..gtf") ->
#   gtf_path
gtf_path = orig_gtf_path

gtf_path %>%
  path_ext_set("protein_coding..gtf") ->
  protein_coding_gtf

gtf_path %>%
  path_ext_set("ncRNA..gtf") ->
  ncrna_gtf

Sys.setenv(GTF=gtf_path)
Sys.setenv(PROTEIN_CODING_GTF=protein_coding_gtf)
Sys.setenv(NCRNA_GTF=ncrna_gtf)

list.files(genome_dir)
```
## Hack GTF to try to make Summary_renaming_and_classification happy
It seems that components of Summary_renaming_and_classification are looking for a ?"gene_name"? entry so lets create a "gene_name" from the gene_id value
```{r}
orig_gtf_path %>%
  read_lines() %>%
  str_replace_all(pattern='gene_id "(.*?)";',
                  replacement='gene_id "\\1"; gene_name "\\1";') %>%
  write_lines(gtf_path)
```

## Prep GTFs


```{r}
orig_gtf_path %>%
  read_lines %>%
  str_subset('gene_biotype "protein_coding"') %>%
  write_lines(protein_coding_gtf)

orig_gtf_path %>%
  read_lines %>%
  str_subset('gene_biotype "ncRNA"') %>%
  write_lines(ncrna_gtf)
```


```{r eval=FALSE, include=FALSE}
gtf_path %>%
  read_tsv(n_max = 20, comment = "#", 
           col_names = c("sequence","source","feature","start","end","score","strand","phase","attributes")) %>%
  filter(str_detect(attributes, 'gene_biotype "protein_coding"')) %>%
  write.table(protein_coding_gtf, sep="\t", quote=FALSE, col.names = FALSE, row.names = FALSE)

gtf_path %>%
  read_tsv(comment = "#", 
           col_names = c("sequence","source","feature","start","end","score","strand","phase","attributes")) %>%
  filter(str_detect(attributes, 'gene_biotype "protein_coding"')) %>%
  write.table(protein_coding_gtf, sep="\t", quote=FALSE, col.names = FALSE, row.names = FALSE)

gtf_path %>%
  read_tsv(comment = "#", 
           col_names = c("sequence","source","feature","start","end","score","strand","phase","attributes")) %>%
  filter(str_detect(attributes, 'gene_biotype "ncRNA"')) %>%
  write.table(ncrna_gtf, sep="\t", quote=FALSE, col.names = FALSE, row.names = FALSE)
```


## OLD Prep GTFs
```{bash eval=FALSE, include=FALSE}
### make protein-coding GTF
awk '$3 == "exon"' $GTF | awk '$19 ~ "protein_coding"' > $PROTEIN_CODING_GTF

### make ncRNA GTF
awk '$3 == "exon"' $GTF | awk '$21 ~ "ncRNA"' > $NCRNA_GTF
```

## Generate Bash script to run stuff in container
### CPAT Setup
http://rna-cpat.sourceforge.net

or use BED file as input
 make_logitModel.py  -x Human_Hexamer.tsv -c Human_coding_transcripts_hg19.bed -n Human_noncoding_transcripts_hg19.bed  -r /database/hg19.fa  -o Human


`make_hexamer_tab.py -c Human_coding_transcripts_CDS.fa   -n Human_noncoding_transcripts_RNA.fa >Human_Hexamer.tsv`

It is unclear if the coding input to make_logitModel.py should be the same as make_hexamer_tab.py
`make_logitModel.py  -x Human_Hexamer.tsv -c Human_coding_transcripts_mRNA.fa -n Human_noncoding_transcripts_RNA.fa -o Human`

```{r}
print(paste("Run", bash_script))
cat(file=bash_script, 
paste0("SCRATCH=", scratch_dir,"\n"),
paste0("LNCPIPE_DIR=", lncpipe_dir,"\n"),
paste0("GENOME_DIR=", genome_dir,"\n"), "

export TMP='/tmp'
export TMPDIR='/tmp'

SINGULARITY_IMAGE='docker://bioinformatist/lncpipe'

singularity exec --bind $GENOME_DIR:$GENOME_DIR $SINGULARITY_IMAGE STAR \\
    --runMode genomeGenerate \\
    --genomeDir $GENOME_DIR \\
    --genomeFastaFiles", genome_path," \\
    --sjdbGTFfile", gtf_path," \\
    --genomeSAindexNbases 11


singularity exec --bind $GENOME_DIR:$GENOME_DIR $SINGULARITY_IMAGE \\
  make_hexamer_tab.py -c", cds_path, "-n", ncrna_path, ">" , hexamer_path,"

singularity exec --bind $GENOME_DIR:$GENOME_DIR $SINGULARITY_IMAGE \\
  make_logitModel.py -c", cds_path, "-n", ncrna_path, "-x", hexamer_path, "-o" , logit_model_prefix, "

git clone https://github.com/granek/LncPipe.git $LNCPIPE_DIR

# lnc_rna/nextflow_1sample.config
# lnc_rna/nextflow_3sample.config
# lnc_rna/nextflow.config
ln -s ", here("lnc_rna/nextflow.config"), "$LNCPIPE_DIR/my_nextflow.config

cd $LNCPIPE_DIR
# git checkout cpat_other_species

nextflow run LncRNAanalysisPipe.nf -c my_nextflow.config -with-singularity $SINGULARITY_IMAGE \\
  --fasta_ref",  genome_path, " \\
  --known_coding_gtf",  protein_coding_gtf, " \\
  --known_lncRNA_gtf",  ncrna_gtf, " \\
  --cpat_hexamer",  hexamer_path, " \\
  --cpat_logit_model",  logit_model_path, " \\
  --out_folder", lncpipe_out_dir, " \\
  --star_index $GENOME_DIR \\
  --input_folder", paste0("'", fastq_link_dir, "/'")," \\
  --fastq_ext '*.fastq.gz'"
)
```


nextflow run LncRNAanalysisPipe.nf -c my_nextflow.config -with-singularity docker://nfcore/lncpipe
nextflow run LncRNAanalysisPipe.nf -c my_nextflow.config -profile standard,singularity

echo $SCRATCH
echo "NEED TO AUTO FIX PATHS IN my_nextflow.config"




