---
title: "Run STAR on TOTAL samples"
output:
  html_document:
    toc: false
---

```{r}
source("run_config.R")

library(dplyr)
library(readr)
library(stringr)
library(GenomicFeatures)
library(Rsamtools)
library(rtracklayer)
library(tibble)
```


# Download and Index Genome
```{bash}
set -u
mkdir -p $GENOME_WITH_MITO_RRNA_DIR
wget --no-verbose --directory-prefix ${GENOME_WITH_MITO_RRNA_DIR} ${FA_URL} 
gunzip -f ${FA_WITH_MITO_RRNA}.gz
```


```{bash}
set -u
mkdir -p ${RRNA_OLIGO_STAROUT}
STAR \
    --runMode genomeGenerate \
    --genomeDir $GENOME_WITH_MITO_RRNA_DIR \
    --genomeFastaFiles ${FA_WITH_MITO_RRNA} \
    --sjdbGTFfile ${GTF_WITH_MITO_RRNA} \
    --outFileNamePrefix ${RRNA_OLIGO_STAROUT}/genome_ \
    --genomeSAindexNbases 11
# --genomeSAindexNbases 6
```

## Check oligos
```{bash}
set -u
SAMPLE="rrna_oligos"
mkdir -p ${RRNA_OLIGO_STAROUT}
    STAR \
    --runMode alignReads \
    --runThreadN $THREADS \
    --genomeDir $GENOME_WITH_MITO_RRNA_DIR \
    --outSAMtype BAM SortedByCoordinate \
    --quantMode GeneCounts \
    --genomeLoad NoSharedMemory \
    --twopassMode None \
    --limitBAMsortRAM 1280000000 \
    --outFileNamePrefix ${RRNA_OLIGO_STAROUT}/${SAMPLE}_ \
    --outFilterScoreMinOverLread 0 --outFilterMatchNminOverLread 0 \
    --outFilterMatchNmin 0 --outFilterMismatchNmax 2 \
    --readFilesIn $RRNA_OLIGOS_FASTQ        
        
    samtools index ${RRNA_OLIGO_STAROUT}/${SAMPLE}_Aligned.sortedByCoord.out.bam
```

## How is STAR counting multimappers
There are some reads from CNAG_10500 that map to an unannotated region between CNAG_10503 and CNAG_03595.  How does STAR count these?  I would like to know for correctly tallying the percentage of reads in total samples that map to rRNA.

### Load STAR table of reads per gene
```{r}
star_oligo_readspergene.path = file.path(rrna_oligo_starout.dir,"rrna_oligos_ReadsPerGene.out.tab")

# load reads per gene count table, dropping rows with all zeroes
read_tsv(star_oligo_readspergene.path, 
         col_names = c("gene_id", "unstranded", "first_strand", "second_strand")) %>%
  filter_at(vars(-gene_id), any_vars(. != 0)) ->
  oligo_read_counts

oligo_read_counts

oligo_read_counts %>%
  summarise_at(vars(-gene_id), sum)
```

### Count number of oligos generated for each rRNA gene
These values should be the same as the the STAR count table above
```{r}
read_tsv(rrna_oligos.file) %>%
  pull(oligo_name) %>%
  str_split("__", simplify = TRUE) %>%
  as_tibble %>%
  dplyr::rename(gene_id="V1", gene_oligo_num="V2") %>%
  group_by(gene_id) %>%
  dplyr::summarise(total_counts = n()) ->
  generated_oligo_counts
  
generated_oligo_counts

generated_oligo_counts %>%
  summarise(total = sum(total_counts))
```

```{r}
full_join(generated_oligo_counts, oligo_read_counts) %>%
  mutate(missing=total_counts-second_strand)
```

### Look for multimappers
```{bash}
samtools view ${RRNA_OLIGO_STAROUT}/rrna_oligos_Aligned.sortedByCoord.out.bam | grep -v NH:i:1 | cut -f1,3-4
```

#### Promiscuos MTrRNA oligos
oligo chrom start other_gene
small_MTrRNA__9	5	478515 CNAG_12438
small_MTrRNA__10	5	478566  CNAG_12438
large_MTrRNA__32	9	71868 CNAG_04124
large_MTrRNA__33	9	71918 CNAG_04124
large_MTrRNA__41	12	483919  CNAG_06164
large_MTrRNA__40	12	483969  CNAG_06164
large_MTrRNA__39	12	484019  CNAG_06164
large_MTrRNA__38	12	484069  CNAG_06164
large_MTrRNA__37	12	484119  CNAG_06164
large_MTrRNA__36	12	484169  CNAG_06164
large_MTrRNA__35	12	484219  CNAG_06164

### Try Counting with RSamtools
https://support.bioconductor.org/p/50774/

#### Load GTF
```{r}
import(gtf_with_mito_rrna.file, format="gtf") %>%
  subset(gene_biotype=="rRNA") %>%
  subset(type=="exon") ->
  rrna_granges
rrna_granges
```

#### Count reads in granges
```{r}
countBam(oligo_bam.file,
         param=ScanBamParam(which=rrna_granges))
```



# SessionInfo
```{r}
sessionInfo()
```
