---
title: "Generate Oligos"
output:
  html_document:
    toc: false
---

# Setup
## Load Libraries
```{r}
# library(Rsamtools)
# library(ggbio)
# library(fs)
# library(GenomicAlignments)
# library(tidyverse)
# library(plyr)
library(rtracklayer)
```

## Load Data
```{r}
source("run_config.R")

h99_genome = readDNAStringSet(fa.file)
# Strip descriptions from chrom names
h99_genome %>%
  names %>%
  str_extract("\\w+") ->
  names(h99_genome)

# 
# seq_dir = "/workspace/2018_tot_samples/seq_out"
# genome_dir = "/workspace/2018_tot_samples/genome/"
# h99_gtf.file = "/workspace/2018_tot_samples/genome/Cryptococcus_neoformans_var_grubii_h99.CNA3.39.gtf"
# 
# 
# 
# 
# h99_gtf.file %>%
#   path_file %>%
#   path_ext_remove %>%
#   paste0("__with_mito_rrna.gtf") %>%
#   file.path(seq_dir, .) ->
#   new_gtf.file
```

```{r eval=FALSE, include=FALSE}
rtrack = import.gff2(gtf_with_mito_rrna.file)
rtrack = rtrack[ rtrack$gene_biotype=="rRNA" ]
rtrack = rtrack[ rtrack$type=="exon" ]
rtrack
```




```{r}
###################################################
### code chunk number 27: makeTxDbFromGFF_1
###################################################
library(GenomicFeatures)
gtf_file <- new_gtf.file
txdb <- makeTxDbFromGFF(gtf_file, format="gtf")
txdb


###################################################
### code chunk number 28: makeTxDbFromGFF_2
###################################################
exonsBy(txdb, by="gene")

```

```{r}
keytypes(txdb)
columns(txdb)
```

