---
title: "Run STAR on TOTAL samples"
output:
  html_document:
    toc: false
---

```{r}
source("run_config.R")
```


# Make adapter file
Don't need to run this chunk because it should have been run by check_rnaseh_oligos.Rmd
```{bash eval=FALSE, include=FALSE}
set -u
mkdir -p $INFO
echo ">Adapter
AGATCGGAAGAGCACACGTCTGAACTCCAGTCA
>AdapterRead2
AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT
>Adapter_rc
TGACTGGAGTTCAGACGTGTGCTCTTCCGATCT
>AdapterRead2_rc
ACACTCTTTCCCTACACGACGCTCTTCCGATCT" >  $ADAPTERS
head $ADAPTERS
```

# Download and Index Genome
Don't need to run this chunk because it should have been run by run_star_on_total_samples.Rmd
```{bash eval=FALSE, include=FALSE}
set -u
mkdir -p $GENOME_WITH_MITO_RRNA_DIR
wget --no-verbose --directory-prefix ${GENOME_WITH_MITO_RRNA_DIR} ${FA_URL} 
gunzip -f ${FA_WITH_MITO_RRNA}.gz
```


```{bash eval=FALSE, include=FALSE}
Don't need to run this chunk because it should have been run by check_rnaseh_oligos.Rmd
set -u
mkdir -p ${ENRICH_COMPARE_STAROUT}
STAR \
    --runMode genomeGenerate \
    --genomeDir $GENOME_WITH_MITO_RRNA_DIR \
    --genomeFastaFiles ${FA_WITH_MITO_RRNA} \
    --sjdbGTFfile ${GTF_WITH_MITO_RRNA} \
    --outFileNamePrefix ${ENRICH_COMPARE_STAROUT}/genome_ \
    --genomeSAindexNbases 11
# --genomeSAindexNbases 6
```


```{bash}
set -u
mkdir -p ${ENRICH_COMPARE_STAROUT}
trim_and_star_func() {
    FASTQ=$1
    FASTQ_BASE=${FASTQ##*/} # strip directory from file path
    SAMPLE="${FASTQ_BASE%_R1_001.fastq.gz}" # strip .fq.gz file extension
    echo $SAMPLE
    echo $FASTQ
    echo $FASTQ_BASE
    # exit 1


    # make a pipe for trimmed fastq
    CUR_PIPE=`mktemp --dry-run`_${SAMPLE}_pipe.fq
    mkfifo $CUR_PIPE

    # Run fastq-mcf
    fastq-mcf \
        $ADAPTERS \
        $FASTQ \
        -o $CUR_PIPE \
        -q 20 -x 0.5 &
        
    STAR \
    --runMode alignReads \
    --runThreadN $THREADS \
    --genomeDir $GENOME_WITH_MITO_RRNA_DIR \
    --outSAMtype BAM SortedByCoordinate \
    --quantMode GeneCounts \
    --genomeLoad NoSharedMemory \
    --twopassMode None \
    --limitBAMsortRAM 1280000000 \
    --outFileNamePrefix ${ENRICH_COMPARE_STAROUT}/${SAMPLE}_ \
    --outFilterScoreMinOverLread 0 --outFilterMatchNminOverLread 0 \
    --outFilterMatchNmin 0 --outFilterMismatchNmax 2 \
    --readFilesIn $CUR_PIPE        
        
    rm -f $CUR_PIPE
    samtools index ${ENRICH_COMPARE_STAROUT}/${SAMPLE}_Aligned.sortedByCoord.out.bam
}
export -f trim_and_star_func

parallel --jobs $MAX_JOBS trim_and_star_func ::: $RAW_FASTQS/[234]_*_S*_L00*_R1_001.fastq.gz
# parallel --jobs $MAX_JOBS trim_and_star_func ::: $RAW_FASTQS/2_TOT*L00[12]*.fastq.gz

```

### Check that custom GTF (with MTrRNA) was used
```{bash}
grep MTrRNA $ENRICH_COMPARE_STAROUT/*_ReadsPerGene.out.tab
```

# SessionInfo
```{r}
sessionInfo()
```
