---
title: "Tally rRNA Reads"
output: html_document
---


```{r}
source("run_config.R")

library(GenomicFeatures)
library(Rsamtools)
library(rtracklayer)

```


# TODO
1. Use RSamtools to count reads mapping to rRNA genes
2. Calculate "total_reads" as sum of "second_strand" column
3. Calculate "strand mapped reads" as sum of "second_strand" column minus "unmapped" reads from that column
4. calculate rRNA_reads/total_reads and rRNA_reads/strand_mapped_reads for each of the 2018 Total libraries

# 

## 1. Use RSamtools to count reads mapping to rRNA genes
https://support.bioconductor.org/p/50774/

### Load GTF
```{r}
import(gtf_with_mito_rrna.file, format="gtf") %>%
  subset(gene_biotype=="rRNA") %>%
  subset(type=="exon") ->
  rrna_granges
rrna_granges
```




### Count reads in granges
```{r}
oligo_bam.file = file.path(mito_rrna_star_out.dir, "rrna_oligos_Aligned.sortedByCoord.out.bam")
countBam(oligo_bam.file,
         param=ScanBamParam(which=rrna_granges)) ->
  rrna_count_tab
rrna_count_tab

# rrna_count_total = 
rrna_count_tab %>%
  pull(records) %>%
  sum ->
  rrna_count_total
```


## 2. Calculate "total_reads" as sum of "second_strand" column

```{r}
read_tsv(star_oligo_readspergene.path, 
         col_names = c("gene_id", "unstranded", "first_strand", "second_strand")) ->
  star_counts
head(star_counts)

# star_counts %>%
#   summarise(total = sum(second_strand))%>%
#   pull(total)
```

```{r}
star_counts %>%
  summarise(total = sum(second_strand))%>%
  pull(total) ->
  total_reads
```

### 3. Calculate "strand mapped reads" as sum of "second_strand" column minus "unmapped" reads from that column

```{r}
star_counts %>%
  filter(gene_id != "N_unmapped") %>%
  summarise(total = sum(second_strand))%>%
  pull(total) ->
  strand_mapped_reads

total_reads
strand_mapped_reads
```

### 4. calculate rRNA_reads/total_reads and rRNA_reads/strand_mapped_reads for each of the 2018 Total libraries
```{r}
cat("rRNA % of Total Reads:", 100 * rrna_count_total/total_reads, fill = TRUE)
cat("rRNA % of Mapped Reads:", 100 * rrna_count_total/strand_mapped_reads, fill = TRUE)

```

#------------------------------------------

```{r}
countsFromBam <- function(bam_file, gene_granges) {
  countBam(bam_file,
           param=ScanBamParam(which=gene_granges)) ->
    count_tab

  count_tab %>%
    pull(records) %>%
    sum %>%
    return
}

summarizeStarCounts <- function(star_count_file) {
  read_tsv(star_count_file, 
         col_names = c("gene_id", "unstranded", "first_strand", "second_strand")) ->
  star_counts
  
  star_counts %>%
    summarise(total = sum(second_strand))%>%
    pull(total) ->
    total_reads
  
  ### 3. Calculate "strand mapped reads" as sum of "second_strand" column minus "unmapped" reads from that column
  
  star_counts %>%
    filter(gene_id != "N_unmapped") %>%
    summarise(total = sum(second_strand))%>%
    pull(total) ->
    mapped_reads
  
  total_reads
  mapped_reads
  return(c(total_reads=total_reads, mapped_reads=mapped_reads))
}

percentRRNA = function(gtf_path, star_out_dir){
  import(gtf_path, format="gtf") %>%
    subset(gene_biotype=="rRNA") %>%
    subset(type=="exon") ->
    rrna_granges
  #-------------------------------------
  bam_suffix = "_Aligned.sortedByCoord.out.bam"
  count_suffix = "_ReadsPerGene.out.tab"
  bamfiles = list.files(star_out_dir, pattern = paste0(bam_suffix,"$"), full.names = TRUE)
  countfiles = bamfiles %>%
    str_replace(bam_suffix, count_suffix)
  #-------------------------------------
  print(bamfiles)
  print(countfiles)
  #-------------------------------------
  # bam_counts = sapply(bamfiles, countsFromBam, rrna_granges)
  # star_counts = sapply(countfiles, summarizeStarCounts)
  # sapply:  as.data.frame(sapply(iris[, 1:4], summary))
  bam_counts = as.data.frame(sapply(bamfiles, countsFromBam, rrna_granges)) %>%
    rownames_to_column  %>%
    mutate(sample=path_file(rowname)) %>%
    mutate(sample=str_replace(sample, bam_suffix,"")) %>%
    dplyr::select(sample, rrna_reads=2 )

  
  star_counts = t(sapply(countfiles, summarizeStarCounts)) %>%
    as.data.frame %>%
    rownames_to_column %>%
    mutate(sample=path_file(rowname)) %>%
    mutate(sample=str_replace(sample, count_suffix,"")) %>%
    dplyr::select(sample, total_reads, mapped_reads)
  
  star_counts
  
  combined_counts = full_join(star_counts, bam_counts, by="sample") %>%
    mutate(rrna_percent_total = 100 * rrna_reads/total_reads,
           rrna_percent_mapped = 100 * rrna_reads/mapped_reads)

  # return(list(bam_counts, star_counts))
  return(combined_counts)

# 1. Use RSamtools to count reads mapping to rRNA genes
# 2. Calculate "total_reads" as sum of "second_strand" column
# 3. Calculate "strand mapped reads" as sum of "second_strand" column minus "unmapped" reads from that column
# 4. calculate rRNA_reads/total_reads and rRNA_reads/mapped_reads for each of the 2018 Total libraries

}
y = percentRRNA(gtf_with_mito_rrna.file, "/workspace/2018_tot_samples/tmp_star_out")
y
```



```{r}
x = percentRRNA(gtf_with_mito_rrna.file, starout.dir)
x

```


```{r}
x2 = x[[2]]
x2[[1]]
```

## 2. Calculate "total_reads" as sum of "second_strand" column



### 4. calculate rRNA_reads/total_reads and rRNA_reads/strand_mapped_reads for each of the 2018 Total libraries
```{r}
cat("rRNA % of Total Reads:", 100 * rrna_count_total/total_reads, fill = TRUE)
cat("rRNA % of Mapped Reads:", 100 * rrna_count_total/strand_mapped_reads, fill = TRUE)

```


