---
title: "Enrichment Correlation"
output:
  
---

```{r global_options, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo=FALSE, include=FALSE)
```

# Load paths and libraries
```{r include=FALSE}
source("run_config_2019.R")
library(foreach)
library(dplyr)
library(tidyr)
library(magrittr)
library(rtracklayer)
library(ggplot2)
library(gridExtra)
library(readr)
library(stringr)
library(Rsamtools)
```

## Load STAR Count Data
```{r}

list.files(enrich_compare_starout.dir, pattern = paste0(count_suffix,"$"), full.names = TRUE) %>%
      path_filter(regex="_2018_P_") ->
  countfiles

countfiles
```

### Load STAR Count tables into a single dataframe
Loading code borrowed from https://gitlab.oit.duke.edu/HTS2018/HTS2018-notebooks/blob/master/pilot/01_Read_Counts.ipynb
```{r}
mycombine <- function(df1, df2) {
    # Combine two data frames by gene names
    #
    # Args:
    #   df1 (Dataframe): the first count data
    #   df2 (Dataframe): the second count data
    #
    # Returns:
    #   (Dataframe) The combined data frame of df1 and df2
    full_join(df1, df2, by = "gene")
}

# Data type for each column
coltypes <- list(col_character(), col_integer(), col_integer(), col_integer())

out <- foreach(count_path = countfiles, .combine = mycombine) %do% {
  # generate rowname (extracted from count file name)
  count_path %>%
    path_file %>%
    str_replace(count_suffix,"") ->
    readset_label 

    # read in the count file
    readr::read_tsv(count_path, col_names = FALSE, col_types = coltypes) %>%
        dplyr::select(X1, X4) %>% # get the 1st and 4th columns (gene ids and second strand read counts)
            dplyr::rename_(.dots=setNames(names(.), c("gene",readset_label)))
}

out[1:6,1:6]

# Drop STAR count statistics and rotate dataframe
out %>%
    dplyr::slice(-(1:4)) %>%
    gather(expid, value, -gene) %>% 
    spread(gene, value) -> genecounts

genecounts[1:6,1:6]
```

### Extract library specific label

```{r}
genecounts %>%
  mutate(Label=str_extract(expid, "^\\d+_\\d+_[A-Z]_[HMT]\\d")) %>%
  dplyr::select(Label, everything()) %>%
  dplyr::select(-expid) ->
  genecounts_with_label
genecounts_with_label

```



### Combine per library counts across lanes
```{r}
read_tsv(metadata.file)
```

```{r}
genecounts_with_label %>%
  group_by(Label) %>%
  summarise_all(funs(sum)) ->
  genecounts_lanes_combined

# Add metadata 
read_tsv(metadata.file) %>%
  dplyr::select(Label, enrichment_method, RNA_sample_num, enrich_rep) ->
  metadata.df

genecounts_lanes_combined %<>%
  left_join(metadata.df, by="Label") %>%
  dplyr::select(library=Label, enrichment_method, RNA_sample_num, enrich_rep, everything())
genecounts_lanes_combined
```

## Compare library preps

### Drop rRNA genes and genes with rRNA homology

#### Find all genes that are mapped by rRNA reads
##### Load GTF
```{r}
import(gtf_with_mito_rrna.file, format="gtf") %>%
  subset(type=="exon") ->
  h99_granges
```



#### Drop genes from count table

```{r}
# genes_to_drop = intersect(colnames(genecounts_lanes_combined), 
#                           strand_specific_rrna_homologs)
genecounts_lanes_combined %>%
  dplyr::select(strand_specific_rrna_homologs) ->
  check_select

check_select[,1:6]

genecounts_lanes_combined %>%
  dplyr::select(-strand_specific_rrna_homologs) ->
  genecounts_lanes_combined_no_rrna

genecounts_lanes_combined_no_rrna[,1:6]

dim(check_select)
dim(genecounts_lanes_combined)
dim(genecounts_lanes_combined_no_rrna)

dplyr::select(genecounts_lanes_combined,small_MTrRNA)
# dplyr::select(genecounts_lanes_combined_no_rrna,small_MTrRNA)

```


### Generate scatterplots for mRNA vs Total and RNaseH vs Total
```{r}
genecounts_lanes_combined_no_rrna[,1:6]

```

```{r}

cur_sample_num = 2
cur_enrich_rep = 1

genecounts_lanes_combined_no_rrna %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

cur_sample.tidy %>%
  summarise(max(mRNA), max(RNaseH)) %>%
  max -> ymax

total_vs_mrna = ggplot(cur_sample.tidy, aes(Total, mRNA)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)
total_vs_rnaseh = ggplot(cur_sample.tidy, aes(Total, RNaseH)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)

grid.arrange(total_vs_mrna, total_vs_rnaseh, nrow = 1)

#------------------------------------------------------------------
mrna_cor = paste("mRNA R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["mRNA"]]) %>%
                 round(2))

rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))
mrna_cor
rnaseh_cor
cat(paste(mrna_cor, rnaseh_cor, sep="\n"))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable, alpha = 1/2)) + 
  geom_point(aes(y=mRNA, color="mRNA"), size=1) +
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  ggtitle(title_text) +
  theme_bw() +
  annotate("text", x=0, y = ymax*0.95, 
           label = paste(mrna_cor, rnaseh_cor, sep="\n"),
           hjust="inward",
           just="inward", 
           color="red")

#------------------------------------------------------------------

ggplot(cur_sample.tidy, alpha = 1/2) +
  geom_point(aes(Total, mRNA), color="red", size=1) + 
  geom_smooth(aes(Total, mRNA), method=lm, se=FALSE, color="red") +
  geom_point(aes(Total, RNaseH), color="blue", size=1) +
  geom_smooth(aes(Total, RNaseH), method=lm, se=FALSE, color="blue") +
  ggtitle(title_text) +
  theme_bw() 


```
```{r correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_no_rrna, cur_sample_num, cur_enrich_rep = 1) {
  
genecounts_lanes_combined_no_rrna %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

cur_sample.tidy %>%
  summarise(max(mRNA), max(RNaseH)) %>%
  max -> ymax

#------------------------------------------------------------------
mrna_cor = paste("mRNA R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["mRNA"]]) %>%
                 round(2))

rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable)) + 
  geom_point(aes(y=mRNA, color="mRNA"), size=1) +
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(mrna_cor, rnaseh_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecounts_lanes_combined_no_rrna, cur_sample, cur_enrich_rep = 1))
}
```






```{r}
# Trying to generalize plotting 
require(dplyr)
set.seed(123)
xx = data.frame(group = rep(1:4, 100), a = rnorm(400) , b = rnorm(400))

xx
gp = group_by(xx, group)
dplyr::summarize(gp, cor(a, b))

# https://www.sixhat.net/how-to-plot-multpile-data-series-with-ggplot.html
```
# testing for all ncRNAs, tRNAs, rRNAs
```{r}
head(h99_granges)
```
```{r}
h99_granges %>% 
  as.data.frame %>% 
  filter(gene_biotype != "protein_coding") %>%
  pull(gene_id) %>%
  unique ->
  all_non_protein_coding_gene_IDs
```
```{r}

all_non_protein_coding_and_rrna_homologs_gene_IDs = 
  union(all_non_protein_coding_gene_IDs, strand_specific_rrna_homologs)

genecounts_lanes_combined %>%
  dplyr::select(all_non_protein_coding_and_rrna_homologs_gene_IDs) ->
  check_select

check_select[,1:6]

genecounts_lanes_combined %>%
  dplyr::select(-all_non_protein_coding_and_rrna_homologs_gene_IDs) ->
  genecounts_lanes_combined_protein_coding_only

genecounts_lanes_combined_protein_coding_only[,1:6]

dim(check_select)
dim(genecounts_lanes_combined)
dim(genecounts_lanes_combined_protein_coding_only)

dplyr::select(genecounts_lanes_combined,small_MTrRNA)

```
# plotting for all ncRNAs, tRNAs, rRNAs
```{r}
cur_sample_num = 2
cur_enrich_rep = 1

genecounts_lanes_combined_protein_coding_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

cur_sample.tidy %>%
  summarise(max(mRNA), max(RNaseH)) %>%
  max -> ymax

total_vs_mrna = ggplot(cur_sample.tidy, aes(Total, mRNA)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)
total_vs_rnaseh = ggplot(cur_sample.tidy, aes(Total, RNaseH)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)

grid.arrange(total_vs_mrna, total_vs_rnaseh, nrow = 1)

#------------------------------------------------------------------
mrna_cor = paste("mRNA R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["mRNA"]]) %>%
                 round(2))

rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))
mrna_cor
rnaseh_cor
cat(paste(mrna_cor, rnaseh_cor, sep="\n"))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable, alpha = 1/2)) + 
  geom_point(aes(y=mRNA, color="mRNA"), size=1) +
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  ggtitle(title_text) +
  theme_bw() +
  annotate("text", x=0, y = ymax*0.95, 
           label = paste(mrna_cor, rnaseh_cor, sep="\n"),
           hjust="inward",
           just="inward", 
           color="red")

#------------------------------------------------------------------

ggplot(cur_sample.tidy, alpha = 1/2) +
  geom_point(aes(Total, mRNA), color="red", size=1) + 
  geom_smooth(aes(Total, mRNA), method=lm, se=FALSE, color="red") +
  geom_point(aes(Total, RNaseH), color="blue", size=1) +
  geom_smooth(aes(Total, RNaseH), method=lm, se=FALSE, color="blue") +
  ggtitle(title_text) +
  theme_bw()

```
```{r protein_coding_only_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_protein_coding_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecounts_lanes_combined_protein_coding_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

cur_sample.tidy %>%
  summarise(max(mRNA), max(RNaseH)) %>%
  max -> ymax

#------------------------------------------------------------------
mrna_cor = paste("mRNA R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["mRNA"]]) %>%
                 round(2))

rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable)) + 
  geom_point(aes(y=mRNA, color="mRNA"), size=1) +
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(mrna_cor, rnaseh_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecounts_lanes_combined_protein_coding_only, cur_sample, cur_enrich_rep = 1))
}
```
```{r}
# Trying to generalize plotting 
require(dplyr)
set.seed(123)
xx = data.frame(group = rep(1:4, 100), a = rnorm(400) , b = rnorm(400))

xx
gp = group_by(xx, group)
dplyr::summarize(gp, cor(a, b))

# https://www.sixhat.net/how-to-plot-multpile-data-series-with-ggplot.html
```

# T2 vs T3 & T4 plotting
```{r}
cur_enrich_method = "Total"
cur_enrich_rep = 1

genecounts_lanes_combined_protein_coding_only %>%
  filter(enrichment_method == cur_enrich_method, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,enrichment_method)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(RNA_sample_num,reads) %>%
  dplyr::rename(Total_2 = `2`, Total_3 = `3`, Total_4 = `4`) ->
  cur_sample.tidy
 

title_text = paste0(cur_enrich_method, collapse=" ")

cur_sample.tidy %>%
  summarise(max(Total_3), max(Total_4)) %>%
  max -> ymax

T2_vs_T3 = ggplot(cur_sample.tidy, aes(Total_2, Total_3)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)

T2_vs_T4 = ggplot(cur_sample.tidy, aes(Total_2, Total_4)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)

grid.arrange(T2_vs_T3, T2_vs_T4, nrow = 1)

#------------------------------------------------------------------
T3_cor = paste("3 R =", 
               cor(cur_sample.tidy[["Total_2"]],
                   cur_sample.tidy[["Total_3"]]) %>%
                 round(2))

T4_cor = paste("4 R =", 
               cor(cur_sample.tidy[["Total_2"]],
                   cur_sample.tidy[["Total_4"]]) %>%
                 round(2))
T3_cor
T4_cor
cat(paste(T3_cor, T4_cor, sep="\n"))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total_2, y=value, color=variable, alpha = 1/2)) + 
  geom_point(aes(y=Total_3, color="T3"), size=1) +
  geom_point(aes(y=Total_4, color="T4"), size=1) +
  ggtitle(title_text) +
  theme_bw() +
  annotate("text", x=0, y = ymax*0.95, 
           label = paste(T3_cor, T4_cor, sep="\n"),
           hjust="inward",
           just="inward", 
           color="red")

#------------------------------------------------------------------

ggplot(cur_sample.tidy, alpha = 1/2) +
  geom_point(aes(Total_2, Total_3), color="red", size=1) + 
  geom_smooth(aes(Total_2, Total_3), method=lm, se=FALSE, color="red") +
  geom_point(aes(Total_2, Total_4), color="blue", size=1) +
  geom_smooth(aes(Total_2, Total_4), method=lm, se=FALSE, color="blue") +
  ggtitle(title_text) +
  theme_bw()

```

```{r correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_protein_coding_only, cur_enrich_method, cur_enrich_rep) {
  
genecounts_lanes_combined_protein_coding_only %>%
  filter(enrichment_method == cur_enrich_method, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,enrichment_method)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(RNA_sample_num,reads) %>%
  dplyr::rename(Total_2 = `2`, Total_3 = `3`, Total_4 = `4`) ->
  cur_sample.tidy
  
title_text = paste0(cur_enrich_method, collapse=" ")

cur_sample.tidy %>%
  summarise(max(Total_3), max(Total_4)) %>%
  max -> ymax

#------------------------------------------------------------------
T3_cor = paste("3 R =", 
               cor(cur_sample.tidy[["Total_2"]],
                   cur_sample.tidy[["Total_3"]]) %>%
                 round(2))

T4_cor = paste("4 R =", 
               cor(cur_sample.tidy[["Total_2"]],
                   cur_sample.tidy[["Total_4"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total_2, y=value, color=variable)) + 
  geom_point(aes(y=Total_3, color="Total_3"), size=1) +
  geom_point(aes(y=Total_4, color="Total_4"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total_2",y="Read Counts", color='Sample') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(T3_cor, T4_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}

  print(makeScatterPlot(genecounts_lanes_combined_protein_coding_only, cur_enrich_method, cur_enrich_rep = 1))
```
```{r correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_protein_coding_only, cur_enrich_method, cur_enrich_rep) {
  
genecounts_lanes_combined_protein_coding_only %>%
  filter(enrichment_method == cur_enrich_method, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,enrichment_method)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(RNA_sample_num,reads) %>%
  dplyr::rename(Total_2 = `2`, Total_3 = `3`, Total_4 = `4`) ->
  cur_sample.tidy
  
title_text = paste0(cur_enrich_method, collapse=" ")

#------------------------------------------------------------------
T3_cor = paste("3 R =", 
               cor(cur_sample.tidy[["Total_2"]],
                   cur_sample.tidy[["Total_3"]]) %>%
                 round(5))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total_2, y=value, color=variable)) + 
  geom_point(aes(y=Total_3, color="Counts"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total_2",y="Read Counts in Total_3", color='Sample') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(T3_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}

  print(makeScatterPlot(genecounts_lanes_combined_protein_coding_only, cur_enrich_method, cur_enrich_rep = 1))
```
```{r correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_protein_coding_only, cur_enrich_method, cur_enrich_rep) {
  
genecounts_lanes_combined_protein_coding_only %>%
  filter(enrichment_method == cur_enrich_method, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,enrichment_method)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(RNA_sample_num,reads) %>%
  dplyr::rename(Total_2 = `2`, Total_3 = `3`, Total_4 = `4`) ->
  cur_sample.tidy
  
title_text = paste0(cur_enrich_method, collapse=" ")

#------------------------------------------------------------------
T4_cor = paste("4 R =", 
               cor(cur_sample.tidy[["Total_2"]],
                   cur_sample.tidy[["Total_4"]]) %>%
                 round(5))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total_2, y=value, color=variable)) + 
  geom_point(aes(y=Total_4, color="Counts"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total_2",y="Read Counts in Total_4", color='Sample') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(T4_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}

  print(makeScatterPlot(genecounts_lanes_combined_protein_coding_only, cur_enrich_method, cur_enrich_rep = 1))
```
# T3 vs T4 plotting
```{r correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_protein_coding_only, cur_enrich_method, cur_enrich_rep) {
  
genecounts_lanes_combined_protein_coding_only %>%
  filter(enrichment_method == cur_enrich_method, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,enrichment_method)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(RNA_sample_num,reads) %>%
  dplyr::rename(Total_2 = `2`, Total_3 = `3`, Total_4 = `4`) ->
  cur_sample.tidy
  
title_text = paste0(cur_enrich_method, collapse=" ")

#------------------------------------------------------------------
T4_cor = paste("4 R =", 
               cor(cur_sample.tidy[["Total_3"]],
                   cur_sample.tidy[["Total_4"]]) %>%
                 round(5))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total_3, y=value, color=variable)) + 
  geom_point(aes(y=Total_4, color="Counts"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total_3",y="Read Counts in Total_4", color='Sample') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(T4_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}

  print(makeScatterPlot(genecounts_lanes_combined_protein_coding_only, cur_enrich_method, cur_enrich_rep = 1))
```
# testing for tRNA correlations
```{r}
h99_granges %>% 
  as.data.frame %>% 
  filter(gene_biotype != "tRNA") %>%
  pull(gene_id) %>%
  unique ->
  all_non_tRNA_gene_IDs
```
```{r}

genecounts_lanes_combined %>%
  dplyr::select(all_non_tRNA_gene_IDs) ->
  check_select

check_select[,1:6]

genecounts_lanes_combined %>%
  dplyr::select(-all_non_tRNA_gene_IDs) ->
  genecounts_lanes_combined_tRNA_only

genecounts_lanes_combined_tRNA_only[,1:6]

dim(check_select)
dim(genecounts_lanes_combined)
dim(genecounts_lanes_combined_tRNA_only)

dplyr::select(genecounts_lanes_combined,small_MTrRNA)

```
# plotting for tRNA correlations
```{r tRNA_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_tRNA_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecounts_lanes_combined_tRNA_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

cur_sample.tidy %>%
  summarise(max(RNaseH), max(mRNA)) %>%
  max -> ymax

#------------------------------------------------------------------
rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))

mrna_cor = paste("mRNA R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["mRNA"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable)) + 
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  geom_point(aes(y=mRNA, color="mRNA"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(rnaseh_cor, mrna_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecounts_lanes_combined_tRNA_only, cur_sample, cur_enrich_rep = 1))
}
```
```{r tRNA_2_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_tRNA_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecounts_lanes_combined_tRNA_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

#------------------------------------------------------------------
rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable)) + 
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y =5, 
           label = paste(rnaseh_cor),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecounts_lanes_combined_tRNA_only, cur_sample, cur_enrich_rep = 1))
}
```
# testing for ncRNA correlations
```{r}
h99_granges %>% 
  as.data.frame %>% 
  filter(gene_biotype != "ncRNA") %>%
  pull(gene_id) %>%
  unique ->
  all_non_ncRNA_gene_IDs
```
```{r}

genecounts_lanes_combined %>%
  dplyr::select(all_non_ncRNA_gene_IDs) ->
  check_select

check_select[,1:6]

genecounts_lanes_combined %>%
  dplyr::select(-all_non_ncRNA_gene_IDs) ->
  genecounts_lanes_combined_ncRNA_only

genecounts_lanes_combined_ncRNA_only[,1:6]

dim(check_select)
dim(genecounts_lanes_combined)
dim(genecounts_lanes_combined_ncRNA_only)

dplyr::select(genecounts_lanes_combined,small_MTrRNA)

```
# plotting for ncRNA correlations
```{r ncRNA_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_ncRNA_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecounts_lanes_combined_ncRNA_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

cur_sample.tidy %>%
  summarise(max(RNaseH), max(mRNA)) %>%
  max -> ymax

#------------------------------------------------------------------
rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))

mrna_cor = paste("mRNA R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["mRNA"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable)) + 
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  geom_point(aes(y=mRNA, color="mRNA"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(rnaseh_cor, mrna_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecounts_lanes_combined_ncRNA_only, cur_sample, cur_enrich_rep = 1))
}
```
```{r ncRNA_2_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecounts_lanes_combined_ncRNA_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecounts_lanes_combined_ncRNA_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

#------------------------------------------------------------------
rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable)) + 
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y =50000, 
           label = paste(rnaseh_cor),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecounts_lanes_combined_ncRNA_only, cur_sample, cur_enrich_rep = 1))
}
```
## Load STAR count data for RZ comparisons
```{r}

list.files(RZ_enrich_samples_starout.dir, pattern = paste0(count_suffix,"$"), full.names = TRUE) %>%
      path_filter(regex="_") ->
  countfiles2018

countfiles2018
```
## Concatenate countfiles and countfilesRZ
```{r}
c(countfiles,countfiles2018) ->
  countfilesRZ

head(countfilesRZ)

length(countfiles)
length(countfiles2018)
length(countfilesRZ)

head(names(countfiles))
head(names(countfiles2018))
head(names(countfilesRZ))

```
### Load STAR Count tables into a single dataframe
Loading code borrowed from https://gitlab.oit.duke.edu/HTS2018/HTS2018-notebooks/blob/master/pilot/01_Read_Counts.ipynb
```{r}
mycombineRZ <- function(df1, df2) {
    # Combine two data frames by gene names
    #
    # Args:
    #   df1 (Dataframe): the first count data
    #   df2 (Dataframe): the second count data
    #
    # Returns:
    #   (Dataframe) The combined data frame of df1 and df2
    full_join(df1, df2, by = "gene")
}

# Data type for each column
coltypesRZ <- list(col_character(), col_integer(), col_integer(), col_integer())

outRZ <- foreach(count_path = countfilesRZ, .combine = mycombineRZ) %do% {
  # generate rowname (extracted from count file name)
  count_path %>%
    path_file %>%
    str_replace(count_suffix,"") ->
    readset_labelRZ 

    # read in the count file
    readr::read_tsv(count_path, col_names = FALSE, col_types = coltypesRZ) %>%
        dplyr::select(X1, X4) %>% # get the 1st and 4th columns (gene ids and second strand read counts)
            dplyr::rename_(.dots=setNames(names(.), c("gene",readset_labelRZ)))
}

outRZ[1:6,1:6]

# Drop STAR count statistics and rotate dataframe
outRZ %>%
    dplyr::slice(-(1:4)) %>%
    gather(expid, value, -gene) %>% 
    spread(gene, value) -> genecountsRZ

genecountsRZ
```
### Extract library specific label

```{r}
genecountsRZ %>%
  mutate(Label=str_extract(expid, "(^\\d+_\\d+_[A-Z]_[HMT]\\d)|(^\\d+_[A-Z]+_[A-Z])")) %>%
  dplyr::select(Label, everything()) %>%
  dplyr::select(-expid) ->
  genecountsRZ_with_label
genecountsRZ_with_label

```
### Combine per library counts across lanes
```{r}
read_tsv(metadata.file)
```

```{r}
genecountsRZ_with_label %>%
  group_by(Label) %>%
  summarise_all(funs(sum)) ->
  genecountsRZ_lanes_combined

# Add metadata 
read_tsv(metadata.file) %>%
  dplyr::select(Label, enrichment_method, RNA_sample_num, enrich_rep) ->
  metadataRZ.df

# need to add enrich_rep to 2018 because 2019 has this since each sample had two technical RNAseH 
read_tsv(metadata_2018.file) %>%
  dplyr::select(Label, enrichment_method, RNA_sample_num) %>%
  mutate(enrich_rep=1) ->
  metadata_2018.df

bind_rows(metadata_2018.df, metadataRZ.df) ->
  metadata_combined.df

genecountsRZ_lanes_combined %<>%
  left_join(metadata_combined.df, by="Label") %>%
  dplyr::select(library=Label, enrichment_method, RNA_sample_num, enrich_rep, everything())
genecountsRZ_lanes_combined
```
## Compare library preps

### Drop rRNA genes and genes with rRNA homology

#### Find all genes that are mapped by rRNA reads
##### Load GTF
```{r}
import(gtf_with_mito_rrna.file, format="gtf") %>%
  subset(type=="exon") ->
  h99_grangesRZ
```
#### Drop genes from count table

```{r}
# genes_to_dropRZ = intersect(colnames(genecountsRZ_lanes_combined), 
#                           strand_specific_rrna_homologs)
genecountsRZ_lanes_combined %>%
  dplyr::select(strand_specific_rrna_homologs) ->
  check_selectRZ

check_selectRZ[,1:6]

genecountsRZ_lanes_combined %>%
  dplyr::select(-strand_specific_rrna_homologs) ->
  genecountsRZ_lanes_combined_no_rrna

genecountsRZ_lanes_combined_no_rrna[,1:6]

dim(check_selectRZ)
dim(genecountsRZ_lanes_combined)
dim(genecountsRZ_lanes_combined_no_rrna)

dplyr::select(genecountsRZ_lanes_combined,small_MTrRNA)
# dplyr::select(genecountsRZ_lanes_combined_no_rrna,small_MTrRNA)

```
# testing for RZ ncRNA correlations
```{r}
h99_grangesRZ %>% 
  as.data.frame %>% 
  filter(gene_biotype != "ncRNA") %>%
  pull(gene_id) %>%
  unique ->
  all_non_ncRNA_gene_IDs_RZ
```
```{r}

genecountsRZ_lanes_combined %>%
  dplyr::select(all_non_ncRNA_gene_IDs_RZ) ->
  check_selectRZ

check_selectRZ[,1:6]

genecountsRZ_lanes_combined %>%
  dplyr::select(-all_non_ncRNA_gene_IDs_RZ) ->
  genecountsRZ_lanes_combined_ncRNA_only

genecountsRZ_lanes_combined_ncRNA_only[,1:6]

dim(check_selectRZ)
dim(genecountsRZ_lanes_combined)
dim(genecountsRZ_lanes_combined_ncRNA_only)

dplyr::select(genecountsRZ_lanes_combined,small_MTrRNA)

```
# plotting for RZ ncRNA correlations - 2018 RZ to 2018 TOT
```{r RZ2018_ncRNA_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecountsRZ_lanes_combined_ncRNA_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecountsRZ_lanes_combined_ncRNA_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

cur_sample.tidy %>%
  summarise(max(RNaseH), max(RZ)) %>%
  max -> ymax

#------------------------------------------------------------------
rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["TOT"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))

RZ_cor = paste("RZ R =", 
               cor(cur_sample.tidy[["TOT"]],
                   cur_sample.tidy[["RZ"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(TOT, y=value, color=variable)) + 
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  geom_point(aes(y=RZ, color="RZ"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(rnaseh_cor, RZ_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecountsRZ_lanes_combined_ncRNA_only, cur_sample, cur_enrich_rep = 1))
}
```
```{r RZ2018_ncRNA_2_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecountsRZ_lanes_combined_ncRNA_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecountsRZ_lanes_combined_ncRNA_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

#------------------------------------------------------------------
RZ_cor = paste("RZ R =", 
               cor(cur_sample.tidy[["TOT"]],
                   cur_sample.tidy[["RZ"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(TOT, y=value, color=variable)) + 
  geom_point(aes(y=RZ, color="RZ"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y =50000, 
           label = paste(RZ_cor),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecountsRZ_lanes_combined_ncRNA_only, cur_sample, cur_enrich_rep = 1))
}
```
# plotting for RZ ncRNA correlations - 2018 RZ to 2019 Total
```{r RZ2019_ncRNA_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecountsRZ_lanes_combined_ncRNA_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecountsRZ_lanes_combined_ncRNA_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

cur_sample.tidy %>%
  summarise(max(RNaseH), max(RZ)) %>%
  max -> ymax

#------------------------------------------------------------------
rnaseh_cor = paste("RNaseH R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RNaseH"]]) %>%
                 round(2))

RZ_cor = paste("RZ R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RZ"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable)) + 
  geom_point(aes(y=RNaseH, color="RNaseH"), size=1) +
  geom_point(aes(y=RZ, color="RZ"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y = ymax, 
           label = paste(rnaseh_cor, RZ_cor, sep="\n"),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecountsRZ_lanes_combined_ncRNA_only, cur_sample, cur_enrich_rep = 1))
}
```
```{r RZ2019_ncRNA_2_correlation_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}

makeScatterPlot = function(genecountsRZ_lanes_combined_ncRNA_only, cur_sample_num, cur_enrich_rep = 1) {
  
genecountsRZ_lanes_combined_ncRNA_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) ->
  cur_sample.tidy
  
title_text = paste0("Sample #", cur_sample_num, collapse=" ")

#------------------------------------------------------------------
RZ_cor = paste("RZ R =", 
               cor(cur_sample.tidy[["Total"]],
                   cur_sample.tidy[["RZ"]]) %>%
                 round(2))
#------------------------------------------------------------------
ggplot(cur_sample.tidy, aes(Total, y=value, color=variable)) + 
  geom_point(aes(y=RZ, color="RZ"), size=1) +
  ggtitle(title_text) +
  labs(x="Read Counts in Total Sample",y="Read Counts in Enriched Sample", color='Enrichment') +
  theme_bw() +
  annotate("text", x=0, y =50000, 
           label = paste(RZ_cor),
           hjust="inward",
           vjust="inward", 
           color="black")
}
for (cur_sample in 2:4) {
  print(makeScatterPlot(genecountsRZ_lanes_combined_ncRNA_only, cur_sample, cur_enrich_rep = 1))
}
```
# setting up 2018 total to 2019 total comparison

```{r 2018_2019_total_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}
head (h99_granges)
head (genecounts_lanes_combined_protein_coding_only)
head (genecountsRZ_lanes_combined_no_rrna)

genecountsRZ

genecountsRZ_lanes_combined

genecountsRZ_lanes_combined %>%
  dplyr::select(-all_non_protein_coding_and_rrna_homologs_gene_IDs) ->
  genecountsRZ_lanes_combined_protein_coding_only

genecountsRZ_lanes_combined_protein_coding_only
```
# Plotting Total 2 comparison
```{r 2018_2019_total_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}
cur_sample_num = "2"
cur_enrich_rep = 1

genecountsRZ_lanes_combined_protein_coding_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) %>%
  dplyr::rename(Total_2_2019 = `Total`, Total_2_2018 = `TOT`,) ->
  cur_sample_total2.tidy
 

title_text = paste0(cur_sample_num, collapse=" ")

cur_sample_total2.tidy %>%
  summarise(max(Total_2_2019), max(Total_2_2018)) %>%
  max -> ymax

T2_2018_vs_T2_2019 = ggplot(cur_sample_total2.tidy, aes(Total_2_2019, Total_2_2018)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)

grid.arrange(T2_2018_vs_T2_2019, nrow = 1)

#------------------------------------------------------------------
T2_cor = paste("2 R =", 
               cor(cur_sample_total2.tidy[["Total_2_2019"]],
                   cur_sample_total2.tidy[["Total_2_2018"]]) %>%
                 round(5))

T2_cor

#------------------------------------------------------------------

ggplot(cur_sample_total2.tidy, alpha = 1/2) +
  geom_point(aes(Total_2_2019, Total_2_2018), color="red", size=1) + 
  geom_smooth(aes(Total_2_2019, Total_2_2018), method=lm, se=FALSE, color="red") +
  ggtitle(title_text) +
  theme_bw() +
  annotate("text", x=0, y = ymax*0.95, 
           label = paste(T2_cor),
           hjust="inward",
           just="inward", 
           color="red")


```
# Plotting Total 3 comparison
```{r 2018_2019_total_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}
cur_sample_num = "3"
cur_enrich_rep = 1

genecountsRZ_lanes_combined_protein_coding_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) %>%
  dplyr::rename(Total_3_2019 = `Total`, Total_3_2018 = `TOT`,) ->
  cur_sample_total3.tidy
 

title_text = paste0(cur_sample_num, collapse=" ")

cur_sample_total3.tidy %>%
  summarise(max(Total_3_2019), max(Total_3_2018)) %>%
  max -> ymax

T3_2018_vs_T3_2019 = ggplot(cur_sample_total3.tidy, aes(Total_3_2019, Total_3_2018)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)

grid.arrange(T3_2018_vs_T3_2019, nrow = 1)

#------------------------------------------------------------------
T3_cor = paste("3 R =", 
               cor(cur_sample_total3.tidy[["Total_3_2019"]],
                   cur_sample_total3.tidy[["Total_3_2018"]]) %>%
                 round(5))

T3_cor

#------------------------------------------------------------------

ggplot(cur_sample_total3.tidy, alpha = 1/2) +
  geom_point(aes(Total_3_2019, Total_3_2018), color="red", size=1) + 
  geom_smooth(aes(Total_3_2019, Total_3_2018), method=lm, se=FALSE, color="red") +
  ggtitle(title_text) +
  theme_bw() +
  annotate("text", x=0, y = ymax*0.95, 
           label = paste(T3_cor),
           hjust="inward",
           just="inward", 
           color="red")

```
# Plotting Total 4 comparison
```{r 2018_2019_total_scatter_plot, include=TRUE, fig.path='figures/', dev=c('png','pdf')}
cur_sample_num = "4"
cur_enrich_rep = 1

genecountsRZ_lanes_combined_protein_coding_only %>%
  filter(RNA_sample_num == cur_sample_num, 
         enrich_rep == cur_enrich_rep) %>%
  dplyr::select(-c(library,RNA_sample_num)) %>%
  gather(gene_id, reads, 2:ncol(.)) %>%
  spread(enrichment_method,reads) %>%
  dplyr::rename(Total_4_2019 = `Total`, Total_4_2018 = `TOT`,) ->
  cur_sample_total4.tidy
 

title_text = paste0(cur_sample_num, collapse=" ")

cur_sample_total4.tidy %>%
  summarise(max(Total_4_2019), max(Total_4_2018)) %>%
  max -> ymax

T4_2018_vs_T4_2019 = ggplot(cur_sample_total4.tidy, aes(Total_4_2019, Total_4_2018)) + 
  geom_point(alpha = 1/2, color="red", size=1) +
  ggtitle(title_text) +
  theme_bw() +
  ylim(0, ymax)

grid.arrange(T4_2018_vs_T4_2019, nrow = 1)

#------------------------------------------------------------------
T4_cor = paste("4 R =", 
               cor(cur_sample_total4.tidy[["Total_4_2019"]],
                   cur_sample_total4.tidy[["Total_4_2018"]]) %>%
                 round(5))

T4_cor

#------------------------------------------------------------------

ggplot(cur_sample_total4.tidy, alpha = 1/2) +
  geom_point(aes(Total_4_2019, Total_4_2018), color="red", size=1) + 
  geom_smooth(aes(Total_4_2019, Total_4_2018), method=lm, se=FALSE, color="red") +
  ggtitle(title_text) +
  theme_bw() +
  annotate("text", x=0, y = ymax*0.95, 
           label = paste(T4_cor),
           hjust="inward",
           just="inward", 
           color="red")

```