---
title: "Check 5'-3' Depth of Lost Genes"
output:
  html_document:
    toc: false
---


```{r global_options, include=FALSE}
# knitr::opts_chunk$set(warning = FALSE, message = FALSE, eval = TRUE, fig.width = 7, fig.height = 4.2)
```

```{r load_libraries, results='hide', warning=FALSE, include=FALSE}
# library(R.utils)
# library(rtracklayer)
library(GenomicFeatures)
# library(stringr)
library(genomation)
library(dplyr)
library(tidyr)
library(ggplot2)
library(readr)

source("../common_config.R")
```

```{r include=TRUE}
warning("Merge Total BAMs")
test_bam = file.path(starout_2019_dir, "SRR12933729_Aligned.sortedByCoord.out.bam")
```

# Plot Depth Across Genes to Check for 3' degradation
## Coverage Depth Function
```{r}
transcriptCoverageAnalysis <- function(gtf_file, bam_path, target_transcript_ids) {
  
  exons = gffToGRanges(gtf_file, filter = "exon")
  exons = exons[exons$gene_biotype=="protein_coding"]
  transcripts = split(exons, exons$transcript_id)
  
  ScoreMatrixBin(target = bam_path, windows = transcripts[target_transcript_ids], bin.num = 100, strand.aware=TRUE) ->
    target_transcript_sm
    
    target_transcript_sm %>%
    t %>%
    as.data.frame %>%
    rename_with(~ target_transcript_ids) %>%
    mutate(mean_coverage = rowMeans(across(where(is.numeric))),
           position_5_3=row_number()) ->
    transcript_sm_df
  
  transcript_sm_df %>%
    ggplot(aes(x=position_5_3,y=mean_coverage)) +
      geom_area(fill="grey50") +
      theme_bw() ->
    average_coverage_plot
    
  transcript_sm_df %>%
    select(-mean_coverage) %>%
    pivot_longer(cols=!position_5_3, names_to="transcript", values_to="depth") ->
    transcript_sm_long
  
  transcript_sm_long %>%
    ggplot(aes(x=position_5_3,y=depth, group=transcript, fill=transcript)) +
      geom_area() +
      facet_grid(rows="transcript", scales = "free_y") +
      scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) +
      theme_bw() +
      theme(strip.text.y = element_text(angle = 0), legend.position="none") ->
    facet_coverage_plot
  
  transcript_sm_long %>%
    ggplot(aes(x=position_5_3, y=transcript, fill=depth)) +
      geom_tile() +
      scale_fill_gradient(low="white", high="blue") +
      scale_x_discrete(expand=c(0,0)) + 
      scale_y_discrete(expand=c(0,0)) +
      theme_bw() ->
    heatmap_coverage_plot
  
  return(list(average_coverage_plot, facet_coverage_plot, heatmap_coverage_plot))
}
```

## Protein Coding Genes Lost in Poly(A) Enrichment
```{r fig.width = 7, fig.height = 7}
gffToGRanges(gtf.file, filter = "exon") %>%
  as_tibble %>%
  select(gene_id, transcript_id) %>%
  distinct %>%
  inner_join(lost_polya_file %>%
               read_csv, 
             by=c("gene_id"="Gene")) ->
  lost_polya_df

lost_polya_df %>%
  filter(`Gene type`== "protein_coding") %>%
  pull(transcript_id) ->
  lost_polya_protein
  
  transcriptCoverageAnalysis(gtf.file, test_bam, lost_polya_protein) ->
  polya_plot_list

polya_plot_list[[1]]
polya_plot_list[[2]]
polya_plot_list[[3]]
```

## Random Protein Coding Genes
```{r fig.width = 7, fig.height = 7}
set.seed(1)

unenriched_count_file %>%
  read_csv %>%
  mutate(total_count = rowSums(across(where(is.numeric)))) ->
  unenriched_counts 

gffToGRanges(gtf.file, filter = "exon") %>%
  as_tibble %>%
  select(gene_id, transcript_id, transcript_biotype) %>%
  distinct %>%
  right_join(unenriched_counts, by="gene_id") %>%
  filter(total_count>=unenriched_low_counts, # get rid of genes with low counts 
         transcript_biotype=="protein_coding") %>% 
  anti_join(lost_polya_df, by="gene_id") %>% # get rid of genes lost in polya
  slice_sample(n=length(lost_polya_protein)) %>%
  pull(transcript_id) %>%
  transcriptCoverageAnalysis(gtf.file, test_bam, .) ->
  random_gene_plot_list

random_gene_plot_list[[1]]
random_gene_plot_list[[2]]
random_gene_plot_list[[3]]
```

## Random Protein Coding Genes With Counts in same range as PolyA Lost genes
```{r fig.width = 7, fig.height = 7}
unenriched_count_file %>%
  read_csv %>%
  mutate(total_count = rowSums(across(where(is.numeric)))) ->
  unenriched_counts 

# Find Total Count Range for Poly-Lost Genes  
left_join(lost_polya_df, unenriched_counts, by="gene_id") %>%
  filter(`Gene type`=="protein_coding") %>%
  summarise(min_count = min(total_count), max_count=max(total_count)) ->
  polya_count_range
  polya_min = pull(polya_count_range, min_count)
  polya_max = pull(polya_count_range, max_count)

gffToGRanges(gtf.file, filter = "exon") %>%
  as_tibble %>%
  select(gene_id, transcript_id, transcript_biotype) %>%
  distinct %>%
  right_join(unenriched_counts, by="gene_id") %>%
  filter(between(total_count, polya_min, polya_max), # only include genes with total counts in same range as the polya-lost genes
         transcript_biotype=="protein_coding") %>% 
  anti_join(lost_polya_df, by="gene_id") %>% # get rid of genes lost in polya
  slice_sample(n=length(lost_polya_protein)) %>%
  pull(transcript_id) %>%
  transcriptCoverageAnalysis(gtf.file, test_bam, .) ->
  random_gene_count_range_plot_list

random_gene_count_range_plot_list[[1]]
random_gene_count_range_plot_list[[2]]
random_gene_count_range_plot_list[[3]]
```

## Work with a Single Gene
#### Plot by Transcripts From GFF - One Gene
```{r}
transcript_list = c("AFR99111")
exons = gffToGRanges(gtf.file, filter = "exon")
transcripts = split(exons, exons$transcript_id)
transcripts_sm = ScoreMatrixBin(target = test_bam, windows = transcripts[transcript_list], bin.num = 100, strand.aware=TRUE)
plotMeta(transcripts_sm)
```

```{r}
gene_list = c("CNAG_06125")
genes = gffToGRanges(gtf.file, filter = "gene")
gene_sm = ScoreMatrixBin(target = test_bam, windows = genes[genes$gene_id %in% gene_list], bin.num = 2316)# , bin.num = 2316, strand.aware=TRUE, bin.op="sum")
plotMeta(gene_sm)
```

# Check Intron Read Depth
Based on above, it seems that ScoreMatrixBin is counting reads that skip introns toward the intron coverage depths (should also be true for any region skipped by a read). 

Let's check this and figure out if there is a way around it
```{r}
library(Rsamtools)
scanBamWhat()
```

```{r}
intronic_gr = GRanges("12", IRanges(376990,377029))

param <- ScanBamParam(which=intronic_gr)

countBam(test_bam, param = param)
# scanBam(test_bam, param = param)
# scanBamWhat()
# bamWhat
```
```{r}
intergenic_gr = GRanges("12", IRanges(378300,378900))

param <- ScanBamParam(which=intergenic_gr, what=c("groupid"))
#                      flag=scanBamFlag(isUnmappedQuery=FALSE))

countBam(test_bam, param = param)
```

# CRUFT Playing with genomation
## Work with a list of transcripts/genes
```{r}
transcript_list = c("AFR99111","AFR98350")
# transcript_list = c("AFR98350")
exons = gffToGRanges(gtf.file, filter = "exon")
transcripts = split(exons, exons$transcript_id)
transcript_sm = ScoreMatrixBin(target = test_bam, windows = transcripts[transcript_list], bin.num = 100, strand.aware=TRUE)
```

### Plot with genomation
```{r}
plotMeta(transcript_sm)
heatMatrix(transcript_sm, col=heat.colors(10))
```

### Plot with GGplot2
```{r}
transcript_sm %>%
  t %>%
  as.data.frame %>%
  rename_with(~ transcript_list) %>%
  mutate(mean_coverage = rowMeans(across(where(is.numeric))),
         position_5_3=row_number()) ->
  transcript_sm_df

transcript_sm_df %>%
  ggplot(aes(x=position_5_3,y=mean_coverage)) +
    geom_line()

transcript_sm_df %>%
  select(-mean_coverage) %>%
  pivot_longer(cols=!position_5_3, names_to="transcript", values_to="depth") %>%
  ggplot(aes(x=position_5_3, y=transcript, fill=depth)) +
    geom_tile() +
    scale_fill_gradientn(colours = heat.colors(10))

warning("Transcripts are mislabelled!!!")
```

# Session Info
```{r}
sessionInfo()
```


