---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Gviz)
library(Rsamtools)
# library(genomation)
library(readr)
library(dplyr)
# library(biomaRt)
options(ucscChromosomeNames=FALSE)

source("../common_config.R")
```

# Combine ALL BAMs
```{r eval=FALSE, include=FALSE}
use_all_bams = TRUE

all_bam_paths = c(list.files(starout_2019_dir, pattern = ".bam$", full.names = TRUE),
                  list.files(starout_2018_dir, pattern = ".bam$", full.names = TRUE))
all_merged_bam_path = file.path(outdir_2019, "merged_all.bam")

if (!file.exists(all_merged_bam_path)){
  mergeBam(all_bam_paths, 
           all_merged_bam_path,
           overwrite=TRUE,
           maxMemory=max_memory_mb)
    indexBam(all_merged_bam_path)
}
```


https://bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.html#49_AlignmentsTrack

```{r eval=FALSE, include=FALSE}
bm <- useMart(host = "grch37.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL", 
              dataset = "hsapiens_gene_ensembl")
# biomTrack <- BiomartGeneRegionTrack(genome = "hg19", chromosome = chr, 
#                                     start = 20e6, end = 21e6,
#                                     name = "ENSEMBL", 
#                                     filter = list(with_refseq_mrna=TRUE), 
#                                     biomart = bm)
# plotTracks(biomTrack, col.line = NULL, col = NULL, stackHeight = 0.3)
```


```{r eval=FALSE, include=FALSE}
afrom <- 2960000
ato <- 3160000
alTrack <- AlignmentsTrack(
  system.file(package = "Gviz", "extdata", "gapped.bam"),
  isPaired = TRUE)
# bmt <- BiomartGeneRegionTrack(genome = "hg19", chromosome = "chr12",
#                               start = afrom, end = ato,
#                               filter = list(with_refseq_mrna = TRUE),
#                               stacking = "dense")
# # We can create a first plot of our read alignments data from a birds eye perspective, say a range of 20,000 nucleotides.
# 
# plotTracks(c(bmt, alTrack), from = afrom, to = ato, 
#            chromosome = "chr12")
```


```{r eval=FALSE, include=FALSE}
alTrack <- AlignmentsTrack(
  system.file(package = "Gviz", "extdata", "gapped.bam"), 
  isPaired = FALSE)
plotTracks(c(alTrack), from = afrom + 12700, to = afrom + 15200, 
           chromosome = "chr12")
```

```{r eval=FALSE, include=FALSE}
afrom <- 44945200
ato <- 44947200
alTrack <- AlignmentsTrack(
  system.file(package = "Gviz", "extdata", "snps.bam"), isPaired = TRUE)
plotTracks(alTrack, chromosome = "chr21", from = afrom, to = ato)
```


```{r eval=FALSE, include=FALSE}
plotTracks(c(alTrack), chromosome = "chr21",
           from = 44946590, to = 44946660)
```

```{r}
alTrack <- AlignmentsTrack(merged_unenriched_2019_bam, isPaired = FALSE)
plotTracks(c(alTrack), from = 798999, to = 801468, 
           chromosome = "1")
```

```{r}
# alTrack <- AlignmentsTrack(merged_unenriched_2019_bam, isPaired = FALSE)
plotTracks(c(alTrack), from = 800462, to = 800543, 
           chromosome = "1")
```


```{r}
read_csv("novel_lncrna.csv") ->
  lncrna_df

lncrna_df
for (row in seq(nrow(lncrna_df))){
  chrom = lncrna_df[[row,"Chromosome"]]
  start = lncrna_df[[row,"Start"]]
  end = lncrna_df[[row,"End"]]
  plotTracks(c(alTrack), from = start, to = end, chromosome = chrom)
}


```


https://bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.html#7_Composite_plots_for_multiple_chromosomes




