---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Gviz)
library(Rsamtools)
# library(genomation)
library(readr)
library(dplyr)
# library(biomaRt)
options(ucscChromosomeNames=FALSE)

source("../common_config.R")
```

# Combine ALL BAMs
```{r}
all_bam_paths = c(list.files(starout_2019_dir, pattern = ".bam$", full.names = TRUE),
                  list.files(starout_2018_dir, pattern = ".bam$", full.names = TRUE))
# all_bam_paths = all_bam_paths[1:3]
all_merged_bam = file.path(outdir_2019, "merged_all.bam")

print(all_merged_bam)
if (!file.exists(all_merged_bam)){
  # samtools merge [-nurlf] [-h inh.sam] [-b <bamlist.fofn>] <out.bam> <in1.bam> [<in2.bam> ... <inN.bam>]
  # samtools merge -f --threads INT
  system2("samtools", args=c("merge",
                             "-f",
                             "--threads", total_threads,
                             all_merged_bam,
                             all_bam_paths))

  system2("samtools", args=c("index",
                             "-@", total_threads,
                             all_merged_bam))
}
```


https://bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.html#49_AlignmentsTrack

GeneRegionTrack



# Test Making Read Plots
```{r}
read_csv("novel_lncrna.csv") ->
  lncrna_df


lncrna_df

lncrna_gtf=here("2019_analysis/final_all.gtf")

grt = GeneRegionTrack(lncrna_gtf)
grt

all_bam_track <- AlignmentsTrack(all_merged_bam, isPaired = FALSE)
chrom = "6"
start = 499352
end = 499840
buf=round((end-start)*0.05)
buf
plotTracks(c(grt, all_bam_track), from = start-buf, to = end+buf, chromosome = chrom, transcriptAnnotation = "gene_id", exonAnnotation= "symbol")
```

# Load Novel lncRNA info
```{r}
read_csv("novel_lncrna.csv") ->
  lncrna_df

lncrna_df %>%
  mutate(igv_loc=paste0(Chromosome, ":", Start, "-", End))
```

# Make Read plots for All Novel lncRNA genes
```{r}
lncrna_gtf=here("2019_analysis/final_all.gtf")

grt = GeneRegionTrack(lncrna_gtf)
all_bam_track <- AlignmentsTrack(all_merged_bam, isPaired = FALSE)

lncrna_df
for (row in seq(nrow(lncrna_df))){
  chrom = lncrna_df[[row,"Chromosome"]]
  start = lncrna_df[[row,"Start"]]
  end = lncrna_df[[row,"End"]]
  # plotTracks(c(all_bam_track), from = start, to = end, chromosome = chrom)
  buf=round((end-start)*0.2)
  plotTracks(c(grt, all_bam_track), from = start-buf, to = end+buf, chromosome = chrom, transcriptAnnotation = "gene_id", exonAnnotation= "symbol")
}
```

https://bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.html#7_Composite_plots_for_multiple_chromosomes

```{r fig.width=8, fig.height=12}  
ncols = 2
nrows = nrow(lncrna_df) %/% ncols
grid.newpage()
pushViewport(viewport(layout = grid.layout(nrows, ncols)))
for (gene_i in seq(nrow(lncrna_df))){
  pushViewport(viewport(layout.pos.col = ((gene_i - 1) %% ncols) + 1,
                        layout.pos.row = (((gene_i) - 1) %/% ncols) + 1))
  chrom = lncrna_df[[gene_i,"Chromosome"]]
  start = lncrna_df[[gene_i,"Start"]]
  end = lncrna_df[[gene_i,"End"]]
  # plotTracks(c(all_bam_track), from = start, to = end, chromosome = chrom)
  buf=round((end-start)*0.2)
  plotTracks(c(grt, all_bam_track), 
             from = start-buf, to = end+buf, 
             chromosome = chrom, 
             transcriptAnnotation = "gene_id", exonAnnotation= "symbol", add=TRUE)
  popViewport(1)
}
```
