---
title: "Tally rRNA Reads"
output: html_document
---

```{r global_options, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo=FALSE, include=FALSE)
```

```{r}
source("run_config_2019.R")

library(GenomicFeatures)
library(Rsamtools)
library(rtracklayer)
library(stringr)
library(tibble)
library(readr)
library(ggplot2)
library(magrittr)
library(rstatix)
library(ggpubr)
```

## Determine rRNA Counts
```{r}
countsFromBam <- function(bam_file, gene_granges) {
  countBam(bam_file,
           param=ScanBamParam(which=gene_granges)) ->
    count_tab

  count_tab %>%
    pull(records) %>%
    sum %>%
    return
}

summarizeStarCounts <- function(star_count_file) {
  read_tsv(star_count_file, 
         col_names = c("gene_id", "unstranded", "first_strand", "second_strand")) ->
    star_counts
  
  star_counts %>%
    summarise(total = sum(second_strand))%>%
    pull(total) ->
    total_reads
  
  ### 3. Calculate "strand mapped reads" as sum of "second_strand" column minus "unmapped" reads from that column
  
  star_counts %>%
    filter(gene_id != "N_unmapped") %>%
    summarise(total = sum(second_strand))%>%
    pull(total) ->
    mapped_reads
  
  # total_reads
  # mapped_reads
  return(c(total_reads=total_reads, mapped_reads=mapped_reads))
}

percentRRNA = function(gtf_path, star_out_dir, bam_subset_regex=".*"){
  import(gtf_path, format="gtf") %>%
    subset(gene_biotype=="rRNA") %>%
    subset(type=="exon") ->
    rrna_granges
  #-------------------------------------
  list.files(star_out_dir, pattern = paste0(bam_suffix,"$"), full.names = TRUE) %>%
    path_filter(regex=bam_subset_regex) ->
    bamfiles

  countfiles = bamfiles %>%
    str_replace(bam_suffix, count_suffix)
  #-------------------------------------
  # print(bamfiles)
  # print(countfiles)
  #-------------------------------------
  bam_counts = as.data.frame(sapply(bamfiles, countsFromBam, rrna_granges)) %>%
    rownames_to_column  %>%
    mutate(sample=path_file(rowname)) %>%
    mutate(sample=str_replace(sample, bam_suffix,"")) %>%
    dplyr::select(sample, rrna_reads=2 )

  
  star_counts = t(sapply(countfiles, summarizeStarCounts)) %>%
    as.data.frame %>%
    rownames_to_column %>%
    mutate(sample=path_file(rowname)) %>%
    mutate(sample=str_replace(sample, count_suffix,"")) %>%
    dplyr::select(sample, total_reads, mapped_reads)
  
  star_counts
  
  combined_counts = full_join(star_counts, bam_counts, by="sample") %>%
    mutate(percent_mapped = 100 * mapped_reads/total_reads,
           rrna_percent_total = 100 * rrna_reads/total_reads,
           rrna_percent_mapped = 100 * rrna_reads/mapped_reads)

  # return(list(bam_counts, star_counts))
  return(combined_counts)

# 1. Use RSamtools to count reads mapping to rRNA genes
# 2. Calculate "total_reads" as sum of "second_strand" column
# 3. Calculate "strand mapped reads" as sum of "second_strand" column minus "unmapped" reads from that column
# 4. calculate rRNA_reads/total_reads and rRNA_reads/mapped_reads for each of the 2018 Total libraries
}
```




```{r}
enrich_compare_counts = percentRRNA(gtf_with_mito_rrna.file, enrich_compare_starout.dir, bam_subset_regex="_2018_P_")
```

## Merge rRNA Counts with Metadata

```{r}
enrich_compare_counts[1:6,1:6]
```


```{r}
metadata.df = read_tsv(metadata.file)

enrich_compare_counts %>%
  mutate(Label=str_extract(sample, "^\\d+_\\d+_[A-Z]_[HMT]\\d"), 
         lane=str_extract(sample, "L\\d{3}$")) %>%
  left_join(metadata.df, by="Label") ->
  enrich_compare_counts_with_meta

kable(enrich_compare_counts_with_meta)
```

## rRNA Reads as a portion of Total Reads
```{r percent_rrna_plot, include=TRUE, fig.width=10, fig.height=5, fig.path='figures/', dev=c('pdf')}

enrich_compare_counts_with_meta %<>%
   mutate(enrichment_method = factor(enrichment_method, 
                                     levels = c("Total", "RNaseH", "mRNA")))

ggplot(data=enrich_compare_counts_with_meta, 
       aes(x=RNA_sample_num, y=rrna_percent_total, fill=lane)) +
  geom_bar(stat="identity", 
           position = position_dodge(preserve = "total"),
           color="black") +
  facet_grid(cols = vars(enrichment_method), scale="free_x") +
  scale_fill_brewer(guide=FALSE) +
  theme_bw() +
  labs(x = "Sample", y="% rRNA reads") +
  ylim(c(0,100)) ->
  percent_rrna.plot

percent_rrna.plot

```

### Prettier Figure for Publication
```{r}
read_csv("rrna_depletion_calculation_2019.csv")
```

#### Summarize Across Lanes
```{r}
enrich_compare_counts_with_meta %>%
  mutate(enrichment_method=str_replace(enrichment_method,pattern="mRNA",replacement="Poly(A)"),
         enrichment_method=factor(enrichment_method, levels = c("Total", "Poly(A)", "RNaseH"))) %>%
  group_by(enrichment_method, RNA_sample_num, sample_year, genotype, condition) %>%
  summarise(
    total_reads=sum(total_reads),
    mapped_reads=sum(mapped_reads),
    rrna_reads=sum(rrna_reads)) %>%
  mutate(ratio_rrna=rrna_reads/total_reads,
         percent_rrna=ratio_rrna*100,
         log_tranform_ratio=log10(ratio_rrna)) ->
  rrna_data_2019
rrna_data_2019
```

#### Make Pretty BarPlot
```{r}
# This chunk modified from plotting_bar_graphs.Rmd

stat.test_2019 <- aov(log_tranform_ratio ~ enrichment_method, data = rrna_data_2019) %>%
  tukey_hsd()
stat.test_2019 

ggbarplot(rrna_data_2019, x = "enrichment_method", y = "percent_rrna", add = c("mean_se"),
          position = position_dodge(),
          color = "black", palette = c("#A7A8A8", "#BBEFFA", "#DFAEC3"),
          fill = "enrichment_method",
          ylim = c(0, 100),
          order = c("Total", "Poly(A)", "RNaseH"),
          legend = "none",
          ylab = "rRNA reads (% total reads)", xlab = " ") +
  stat_pvalue_manual(
    stat.test_2019, label = "p.adj.signif", 
    tip.length = 0,
    size = 6,
    bracket.size = 0.75,
    y.position = c(95, 100, 8.5))
```

## Mapped Reads as a portion of Total Reads
```{r}
ggplot(data=enrich_compare_counts_with_meta, 
       aes(x=RNA_sample_num, y=percent_mapped, fill=lane)) +
  geom_bar(stat="identity", 
           position = position_dodge(preserve = "total"),
           color="black") +
  facet_grid(cols = vars(enrichment_method), scale="free_x") +
  scale_fill_brewer(guide=FALSE) +
  theme_bw() +
  labs(x = "Sample", y="% Reads Mapped") +
  ylim(c(0,100))

```

