---
title: "Check 5'-3' Depth of Lost Genes"
output:
  html_document:
    toc: false
---


```{r global_options, include=FALSE}
# knitr::opts_chunk$set(warning = FALSE, message = FALSE, eval = TRUE, fig.width = 7, fig.height = 4.2)
```

```{r load_libraries, results='hide', warning=FALSE, include=FALSE}
# library(R.utils)
# library(rtracklayer)
library(GenomicFeatures)
# library(stringr)
library(genomation)
library(dplyr)
library(tidyr)
library(ggplot2)
library(readr)
library(Rsamtools)


source("../common_config.R")

use_all_bams=TRUE
```

# Merge Unenriched BAMs
```{r echo=FALSE}
accessions_file %>%
  read_csv(col_types = "cccccccii") %>%
  filter(sequencing_batch=="2019",
         enrichment_method=="Unenriched") %>%
  pull(Run) %>%
  paste0("_Aligned.sortedByCoord.out.bam") %>%
  file.path(starout_2019_dir,.) ->
  unenriched_2019_bam_paths

if (!file.exists(merged_unenriched_2019_bam)){
  # samtools merge [-nurlf] [-h inh.sam] [-b <bamlist.fofn>] <out.bam> <in1.bam> [<in2.bam> ... <inN.bam>]
  # samtools merge -f --threads INT
  system2("samtools", args=c("merge",
                             "-f",
                             "--threads", total_threads,
                             merged_unenriched_2019_bam,
                             unenriched_2019_bam_paths))

  system2("samtools", args=c("index",
                             "-@", total_threads,
                             merged_unenriched_2019_bam))
}
```

# Plot Depth Across Genes to Check for 3' degradation
```{r CoverageDepthFunction, echo=FALSE}
transcriptCoverageAnalysis <- function(gtf_file, bam_path, target_transcript_ids) {
  
  exons = gffToGRanges(gtf_file, filter = "exon")
  exons = exons[exons$gene_biotype=="protein_coding"]
  transcripts = split(exons, exons$transcript_id)
  
  ScoreMatrixBin(target = bam_path, windows = transcripts[target_transcript_ids], bin.num = 100, strand.aware=TRUE) ->
    target_transcript_sm
    
    target_transcript_sm %>%
    t %>%
    as.data.frame %>%
    rename_with(~ target_transcript_ids) %>%
    mutate(mean_coverage = rowMeans(across(where(is.numeric))),
           position_5_3=row_number()) ->
    transcript_sm_df
  
  transcript_sm_df %>%
    ggplot(aes(x=position_5_3,y=mean_coverage)) +
      geom_area(fill="grey50") +
      theme_bw() +
      labs(x ="Position Along Transcript", y="Average Read Depth") +
      scale_x_discrete(expand=c(0,0)) ->
    average_coverage_plot
    
  transcript_sm_df %>%
    select(-mean_coverage) %>%
    pivot_longer(cols=!position_5_3, names_to="transcript", values_to="depth") ->
    transcript_sm_long
  
  transcript_sm_long %>%
    ggplot(aes(x=position_5_3,y=depth, group=transcript, fill=transcript)) +
      geom_area() +
      facet_grid(rows="transcript", scales = "free_y") +
      scale_y_continuous(breaks = scales::pretty_breaks(n = 2)) +
      theme_bw() +
      theme(strip.text.y = element_text(angle = 0), legend.position="none") +
      labs(x ="Position Along Transcript", y="Read Depth") +
      scale_x_discrete(breaks = c(10,90),
                       labels = c("5'","3'")) ->
    facet_coverage_plot
  
  transcript_sm_long %>%
    ggplot(aes(x=position_5_3, y=transcript, fill=depth)) +
      geom_tile() +
      scale_fill_gradient(low="white", high="blue") +
      scale_x_discrete(breaks = c(1,100),
                       labels = c("5'","3'")) + 
      labs(x ="Position Along Transcript", x ="Transcript ID", fill = "Read Depth") +
      theme_bw() ->
    heatmap_coverage_plot
  
  return(list(average_coverage_plot, facet_coverage_plot, heatmap_coverage_plot))
}
# scale_x_discrete(breaks = c(10,90),
#                  labels = c("5'","3'"),
#                  expand=c(0,0)) ->
```

## Protein Coding Genes Lost in Poly(A) Enrichment
We identified a small number of genes that are expressed, as identified in unenriched libraries and libraries enriched with Ribo-Zero and the RNase H method. However these genes are "lost" by the Poly(A) enrichment method - these genes have very few counts in libraries enriched by the Poly(A) method. About half of these genes are annotated as ncRNA, and are therefore expected to be lost by the Poly(A) method because, in general, only protein coding transcripts are polyadenylated. The rest of the genes encode proteins; one possible explanation for loss of these transcripts by the Poly(A) method is that these transcripts are being rapidly turned over, and therefore the Poly(A) tails are degraded. To investigate this possibility, we examined the read coverage across these transcripts. If these transcripts were being rapidly degraded, we would expect a 5' bias in transcript read coverage (i.e. increasing read depth toward the 5' end), since mRNA is degraded from the 3' end [REF]. Note that in the figures transcript lengths are normalized, so the left end of the the plot corresponds to th 5' end of each transcript and the right end corresponds to the 3' end. Also note that only predicted transcripts are shown, so introns are not show.

Read depth across transcripts is visualized in three ways - a plot of the average depth across all lost protein coding transcripts, and each transcript visualized individually as a filled area plot and a heatmap. While transcript AFR92445 (CNAG_00311) shows a strong peak at its 5'-end, the other transcripts do not appear to have any 5'-bias, suggesting that partial transcript degradation is not the cause of loss of these transcripts by the Poly(A) method. 

```{r echo=FALSE, fig.width = 7, fig.height = 7}
gffToGRanges(gtf.file, filter = "exon") %>%
  as_tibble %>%
  dplyr::select(gene_id, transcript_id) %>%
  distinct %>%
  inner_join(lost_polya_file %>%
               read_csv(col_types="ccc"), 
             by=c("gene_id"="Gene")) ->
  lost_polya_df

lost_polya_df %>%
  filter(`Gene type`== "protein_coding") %>%
  pull(transcript_id) ->
  lost_polya_protein
  
  transcriptCoverageAnalysis(gtf.file, merged_unenriched_2019_bam, lost_polya_protein) ->
  polya_plot_list

polya_plot_list[[1]]
polya_plot_list[[2]]
polya_plot_list[[3]]
```
### Drop Mito Genes

```{r echo=FALSE, fig.width = 7, fig.height = 3}
gffToGRanges(gtf.file, filter = "exon") %>%
  as_tibble %>%
  select(gene_id, transcript_id) %>%
  distinct %>%
  inner_join(lost_polya_file %>%
               read_csv(col_types="ccc"), 
             by=c("gene_id"="Gene")) %>%
  filter(Chromosome != "Mt") ->
  lost_polya_df

lost_polya_df

lost_polya_df %>%
  filter(`Gene type`== "protein_coding") %>%
  pull(transcript_id) ->
  lost_polya_protein
  
  transcriptCoverageAnalysis(gtf.file, merged_unenriched_2019_bam, lost_polya_protein) ->
  polya_plot_list

polya_plot_list[[1]]
polya_plot_list[[2]]
polya_plot_list[[3]]
```




## Control 1: Random Protein Coding Genes
As a control we similarly examined the read depth across the transcripts of a random sample of protein coding genes. Genes were excluded that were in the "lost" gene set or had low counts (fewer than `r unenriched_low_counts` total reads across the unenriched samples). Compared with "lost" transcript there are no obvious differences, except the lost transcripts have much higher expression. Therefore we re-ran this analysis with a random sample of genes that more similar expression to the lost genes (genes were sampled that had total counts across the unenriched libraries in the same range as the lost genes). Examination found the same results.
```{r echo=FALSE, fig.width = 7, fig.height = 3}
set.seed(1)

unenriched_count_file %>%
  read_csv(col_types = "ciiiiii") %>%
  mutate(total_count = rowSums(across(where(is.numeric)))) ->
  unenriched_counts 

gffToGRanges(gtf.file, filter = "exon") %>%
  as_tibble %>%
  select(gene_id, transcript_id, transcript_biotype) %>%
  distinct %>%
  right_join(unenriched_counts, by="gene_id") %>%
  filter(total_count>=unenriched_low_counts, # get rid of genes with low counts 
         transcript_biotype=="protein_coding") %>% 
  anti_join(lost_polya_df, by="gene_id") %>% # get rid of genes lost in polya
  slice_sample(n=length(lost_polya_protein)) %>%
  pull(transcript_id) %>%
  transcriptCoverageAnalysis(gtf.file, merged_unenriched_2019_bam, .) ->
  random_gene_plot_list

random_gene_plot_list[[1]]
random_gene_plot_list[[2]]
random_gene_plot_list[[3]]
```

## Control 2: Random Protein Coding Genes with similar count range as PolyA Lost genes (quantile)
```{r echo=FALSE, fig.width = 7, fig.height = 3}
set.seed(1)

unenriched_count_file %>%
  read_csv(col_types = "ciiiiii") %>%
  mutate(total_count = rowSums(across(where(is.numeric)))) ->
  unenriched_counts 

# Find Total Count Range for Poly-Lost Genes  
left_join(lost_polya_df, unenriched_counts, by="gene_id") %>%
  filter(`Gene type`=="protein_coding") %>%
  pull(total_count) ->
  total_counts
  quant_25 = quantile(total_counts, type=1)[2] 
  quant_75 = quantile(total_counts, type=1)[4] 

gffToGRanges(gtf.file, filter = "exon") %>%
  as_tibble %>%
  select(gene_id, transcript_id, transcript_biotype) %>%
  distinct %>%
  right_join(unenriched_counts, by="gene_id") %>%
  filter(between(total_count, quant_25, quant_75), # only include genes with total counts in same range as the polya-lost genes
         transcript_biotype=="protein_coding") %>% 
  anti_join(lost_polya_df, by="gene_id") %>% # get rid of genes lost in polya
  slice_sample(n=length(lost_polya_protein)) %>%
  pull(transcript_id) %>%
  transcriptCoverageAnalysis(gtf.file, merged_unenriched_2019_bam, .) ->
  random_gene_count_range_plot_list

random_gene_count_range_plot_list[[1]]
random_gene_count_range_plot_list[[2]]
random_gene_count_range_plot_list[[3]]
```

## Work with a Single Gene
#### Plot by Transcripts From GFF - One Gene
```{r eval=FALSE, include=FALSE}
transcript_list = c("AFR99111")
exons = gffToGRanges(gtf.file, filter = "exon")
transcripts = split(exons, exons$transcript_id)
transcripts_sm = ScoreMatrixBin(target = single_bam, windows = transcripts[transcript_list], bin.num = 100, strand.aware=TRUE)
plotMeta(transcripts_sm)
```

```{r eval=FALSE, include=FALSE}
gene_list = c("CNAG_06125")
genes = gffToGRanges(gtf.file, filter = "gene")
gene_sm = ScoreMatrixBin(target = single_bam, windows = genes[genes$gene_id %in% gene_list], bin.num = 2316)# , bin.num = 2316, strand.aware=TRUE, bin.op="sum")
plotMeta(gene_sm)
```

# Testing Things Out
## Check Intron Read Depth
Based on above, it seems that ScoreMatrixBin is counting reads that skip introns toward the intron coverage depths (should also be true for any region skipped by a read). 

Let's check this and figure out if there is a way around it
```{r eval=FALSE, include=FALSE}
library(Rsamtools)
scanBamWhat()
```

```{r eval=FALSE, include=FALSE}
intronic_gr = GRanges("12", IRanges(376990,377029))

param <- ScanBamParam(which=intronic_gr)

countBam(single_bam, param = param)
# scanBam(single_bam, param = param)
# scanBamWhat()
# bamWhat
```
```{r eval=FALSE, include=FALSE}
intergenic_gr = GRanges("12", IRanges(378300,378900))

param <- ScanBamParam(which=intergenic_gr, what=c("groupid"))
#                      flag=scanBamFlag(isUnmappedQuery=FALSE))

countBam(single_bam, param = param)
```

# Session Info
```{r}
sessionInfo()
```


