---
title: "Enrichment Correlation"
output: html_document
---

```{r global_options, include=FALSE, echo=FALSE}
library(knitr)
knitr::opts_chunk$set(echo=TRUE, include=TRUE)
```

# Load paths and libraries
```{r include=FALSE}
source("shared_functions.R")
source("combined_config.R")

library(dplyr)
library(ggplot2)
library(DESeq2)
library(tibble)
library(cowplot)
library(rtracklayer)
library(here)

fig_palette="Dark2"
```

# Analysis with Normalized Counts
## Load STAR Count Data
```{r}
starout_2019_dir %>%
  list.files(pattern = paste0("_2018_.*",count_suffix,"$"), full.names = TRUE) %>%
  loadStarCounts ->
  genecounts_2019_run

starout_2018_dir %>%
  list.files(pattern = paste0(count_suffix,"$"), full.names = TRUE) %>%
  loadStarCounts ->
  genecounts_2018_run
```

## Load Data in DESeq2
### Make Metadata dataframe for DESeq
```{r}
genecounts_2018_run %>%
  dplyr::select(expid) %>%
  dplyr::mutate(library=str_extract(expid, "^\\d+_[A-Z]+_[A-Z]"),
         exp_year=2018,
         enrichment_rep="1") %>%
  tidyr::separate(library, 
           into = c("RNA_sample_num", "enrichment_method",NA),
           sep="_", remove=FALSE) ->
  sample_df_2018

# knitr::kable(sample_df_2018)
```

```{r}
genecounts_2019_run %>%
  select(expid) %>%
  mutate(library=str_extract(expid, "^\\d+_2018_[A-Z]+_[HMT]\\d"),
         exp_year=2019) %>%
  separate(library, 
           into = c("RNA_sample_num", NA, NA,"method_rep"),
           sep="_", remove=FALSE) %>%
  separate(method_rep, 
           into = c("enrichment_method", "enrichment_rep"),
           sep=1) ->
  sample_df_2019
knitr::kable(sample_df_2019)
 

bind_rows(sample_df_2018, sample_df_2019) %>%
  mutate(enrichment_method = case_when(enrichment_method == "H" ~ "RNase H",
                                       enrichment_method %in% c("T","TOT") ~ "Unenriched",
                                       enrichment_method %in% c("M","MA") ~ "Poly(A)",
                                       enrichment_method == "RZ" ~ "Ribo-Zero")) %>%
  mutate(RNA_sample_num = case_when(RNA_sample_num == "2" ~ "A",
                                    RNA_sample_num == "3" ~ "B",
                                    RNA_sample_num == "4" ~ "C")) %>%
  mutate(enrichment_method=factor(enrichment_method, 
                                  levels = c("Unenriched", "RNase H", "Ribo-Zero", "Poly(A)"))) %>%
  column_to_rownames(var="expid") ->
  sample_df
# knitr::kable(sample_df)
```

### Make Count Matrix for DESeq
```{r}
bind_rows(genecounts_2018_run, genecounts_2019_run) %>%
  # drop rRNA and rRNA homologs
  select(-all_of(strand_specific_rrna_homologs)) %>%
  column_to_rownames(var="expid") %>%
  as.matrix %>%
  t ->
  genecounts_for_desesq
# genecounts_for_desesq[1:6,1:6]
```

### Make DESeq object
```{r}
stopifnot(all(rownames(sample_df) == colnames(genecounts_for_desesq))) # Confirm that sample metadata and count columns are in the same order

DESeqDataSetFromMatrix(countData = genecounts_for_desesq,
                       colData = sample_df,
                       design = ~ enrichment_method) ->
  dds
colData(dds)

# Collapse Replicates
dds %>% collapseReplicates(object=., 
                           groupby=colData(.) %>% 
                             as_tibble %>%
                             pull(library)) ->
  dds

# extract levels to use for color palettes
dds %>%
  colData %>%
  as_tibble %>%
  pull(enrichment_method) %>%
  levels ->
  enrich_levels

dds
```

```{r}
dds <- DESeq(dds)
```

## Calculate Group Mean of Normalized counts

```{r}
dds %>%
  counts(normalized=TRUE) ->
  norm_counts

norm_counts %>%
  as_tibble(rownames = "geneid") %>%
  mutate(gold_mean = rowMeans(select(., matches("_TOT_|_T1")))) ->
  norm_counts_wmeans
```

# Correlations of all Libraries with Average Unenriched
## Pearson R Correlation

```{r}
enrichedCorrelationToGoldStandard <- function(norm_counts_wmeans, dds) {
  norm_counts_wmeans %>%
    select(-geneid) %>%
    summarise_all(.funs=cor, y=pull(., gold_mean)) %>%
    t %>%
    as_tibble(rownames = "library", .name_repair="universal") %>%
    dplyr::rename(cor_with_unenriched="...1") ->
    enriched_cor
  
  dds %>%
    colData %>%
    as_tibble %>%
    left_join(enriched_cor, by="library") %>%
    filter(enrichment_method!="Unenriched") ->
  enriched_cor
  return(enriched_cor)
}
```


Pearson correlation of each library with the "Gold Standard".

For each library per gene read counts were normalized by that library's total reads mapped to annotated genes (excluding rRNA genes and genes that recruit rRNA reads). An "gold standard" per gene read count was calculated as the average normalized count across the six unenriched libraries.  The Pearson correlation was then calculated between this gold standard and the normalized gene counts for each library.  All unenriched libraries are highly correlated with the gold standard unenriched, indicating that biological and technical replicates are all very similar.  The RNase H libraries are better correlated with the gold standard than Ribo Zero libraries and Poly A libraries, with Poly A libraries having the worst correlation.

## Leave-one-out Correlation for Unenriched
do "cross validation" mean
calculate mean of leave one out set, then correlate the left out with with mean of other 5

### Calculate Cross Correlations
```{r}
norm_counts %>%
  as_tibble(rownames = "geneid") %>%
  select(geneid, matches("_TOT_|_T1")) ->
  unenriched_norm_counts
```

```{r}
calculateCrossCorrelation <- function(unenriched_norm_counts, dds) {
  data.frame(row.names = colnames(unenriched_norm_counts),
         cor_with_unenriched = rep(-1, ncol(unenriched_norm_counts))) ->
    unenriched_cor
  
  for (cur_sample in colnames(unenriched_norm_counts)){
    # print(cur_sample)
    unenriched_norm_counts %>%
      select(-cur_sample) %>%
      rowMeans %>%
      cor(select(unenriched_norm_counts, all_of(cur_sample))) ->
      unenriched_cor[cur_sample,"cor_with_unenriched"]
  }
  unenriched_cor %>%
    as_tibble(rownames = "library", .name_repair="universal") ->
    unenriched_cor
  
  dds %>%
    colData %>%
    as_tibble %>%
    right_join(unenriched_cor, by="library") ->
    unenriched_cor
  
  return(unenriched_cor)
}
```

### Plot Correlations
```{r}
# Find range of R values so plots have the same y-range
plotCorrelationValues <- function(unenriched_cor, enriched_cor, dds, fig_palette, enrich_levels) {
  bind_rows(unenriched_cor, enriched_cor) %>%
    pull(cor_with_unenriched) %>%
    range ->
    r_range
  
  set.seed(3)
  unenriched_cor %>%
    ggplot(aes(x=enrichment_method, y=cor_with_unenriched, color=enrichment_method)) +
    geom_boxplot(fill=NA, color="grey70",outlier.color = NA, width=0.5) +
    geom_jitter(width=0.2, height=0, size = 4, stroke = 1, alpha=0.5) +
    scale_colour_brewer(palette = fig_palette, guide = FALSE, drop=TRUE, limits = enrich_levels) +
    labs(x="Enrichment Method", y="Correlation") +
    theme_minimal() +
    theme(text=element_text(size=10)) +
    ylim(r_range) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) ->
    unenriched_plot
  
  set.seed(3)
  enriched_cor %>%  
    ggplot(aes(x=enrichment_method, y=cor_with_unenriched, color=enrichment_method)) +
    geom_boxplot(fill=NA, color="grey70",outlier.color = NA, width=0.5) +
    geom_jitter(width=0.2, height=0, size = 4, stroke = 1, alpha=0.5) +
    scale_colour_brewer(palette = fig_palette, guide = FALSE, drop=TRUE, limits = enrich_levels) +
    labs(x="Enrichment Method", y="Correlation") +
    theme_minimal() +
    theme(text=element_text(size=10)) +
    ylim(r_range) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) ->
    enriched_plot
  
  plot_grid(unenriched_plot, enriched_plot, 
            rel_widths = c(1, 2),
            labels = c('A', 'B'), 
            label_size = 10)
}
```

```{r}
correlationRanges <- function(unenriched_cor, enriched_cor, rounding=3){
  bind_rows(unenriched_cor, enriched_cor) %>%
  group_by(enrichment_method) %>%
  summarise(min_r = round(min(cor_with_unenriched), rounding),
            max_r = round(max(cor_with_unenriched), rounding)) %>%
    return
}
```


TODO:
1. Make correlation and scatter plots for ncRNA and protein_coding
  A. wrap plotting in function(s)

# Supplementary Scatter Facet
## Prep data for plotting
```{r}

norm_counts_wmeans %>%
  pivot_longer(
    cols = c(-geneid, -gold_mean),
    names_to = "library",
    values_to = "norm_count") ->
  norm_counts_long

dds %>%
  colData %>%
  as_tibble %>%
  select(-sizeFactor) %>%
  left_join(norm_counts_long, by="library") ->
  norm_counts_long
# knitr::kable(norm_counts_long)

norm_counts_long %>%
    # only plot one tech rep per library: remove RNAseH rep2 and Poly A and Unenriched from 2018
    filter(enrichment_rep != 2) %>%
    filter(!(enrichment_method %in% c("Poly(A)","Unenriched") &
               exp_year == 2018)) %>%
    filter(enrichment_method!="Unenriched") ->
  norm_counts_long_noreps
```

## Check technical replicates
```{r}
norm_counts_long %>%
  filter(enrichment_rep != 2) %>%
  filter(!(enrichment_method %in% c("Poly(A)","Unenriched") &
             exp_year == 2018)) %>%
  select(RNA_sample_num, enrichment_method, exp_year, enrichment_rep) %>%
  table
```

## Make Sample x Enrichment Facetted Scatter Plot

```{r}
makeEnrichmentScatterplot <- function(norm_counts_long, fig_palette, enrich_levels) {
  norm_counts_long %>%
    ggplot(aes(x=gold_mean, y=norm_count, color=enrichment_method)) +
    geom_point(alpha=0.3) +
    labs(x = "Average Normalized Count in Unenriched", y= "Normalized Count") +
    theme_bw() +
    theme(text=element_text(size=10)) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())  +
    facet_grid(enrichment_method ~ RNA_sample_num) +
    scale_colour_brewer(palette = fig_palette, guide = FALSE, drop=TRUE, limits = enrich_levels) +
    geom_abline(size=0.5) %>%
    return
}
```


# Load GTF info
```{r}
import(gtf_with_mito_rrna.file, format="gtf") %>%
  subset(type=="exon") %>% 
  as.data.frame ->
  h99_granges
```

# All Plots
## All Genes
```{r}
unenriched_norm_counts %>%
  select(-geneid) %>%
  calculateCrossCorrelation(dds) ->
  unenriched_cor

enriched_cor = enrichedCorrelationToGoldStandard(norm_counts_wmeans, dds)

correlationRanges(unenriched_cor, enriched_cor) %>%
  kable
```

```{r}
plotCorrelationValues(unenriched_cor, enriched_cor, dds, fig_palette, enrich_levels)
ggsave(file.path(fig.dir, "Figure_2.tiff"), compression = "lzw", width=15, height=10.71, units = "cm")
```

```{r fig.width=8, fig.height=8}  
makeEnrichmentScatterplot(norm_counts_long_noreps, fig_palette, enrich_levels)
ggsave(file.path(fig.dir, "Figure_S1.tiff"), compression = "lzw", width=20, height=20, units = "cm")
```

For each library per gene read counts were normalized by that library's total reads mapped to annotated genes (excluding rRNA genes and genes that recruit rRNA reads). A "gold standard" per gene read count was calculated as the average normalized count across the six unenriched libraries.  Libraries for each enrichment method applied to each biological replicate are ploted (technical replicates are not shown).  Each point represents a gene with the gold standard count as the x value, and the normalized count from a library as the y value.


## Protein Coding Genes

```{r}
h99_granges %>% 
  filter(gene_biotype == "protein_coding") %>%
  pull(gene_id) %>%
  unique ->
  protein_coding_geneids

unenriched_norm_counts %>%
  filter(geneid %in% protein_coding_geneids) %>%
  select(-geneid) %>%
  calculateCrossCorrelation(dds) ->
  protein_coding_unenriched_cor

norm_counts_wmeans %>%
  filter(geneid %in% protein_coding_geneids) %>%
  enrichedCorrelationToGoldStandard(dds) ->
  protein_coding_enriched_cor

correlationRanges(protein_coding_unenriched_cor, protein_coding_enriched_cor) %>%
  kable

```

```{r}
plotCorrelationValues(protein_coding_unenriched_cor, 
                      protein_coding_enriched_cor, 
                      dds, fig_palette, enrich_levels)
ggsave(file.path(fig.dir, "Figure_3.tiff"), compression = "lzw", width=15, height=10.71, units = "cm")
```

```{r fig.width=8, fig.height=8}  
norm_counts_long_noreps %>%
  filter(geneid %in% protein_coding_geneids) %>%
  makeEnrichmentScatterplot(fig_palette, enrich_levels)
ggsave(file.path(fig.dir, "Figure_S2.tiff"), compression = "lzw", width=20, height=20, units = "cm")
```

## ncRNA Genes
```{r}
h99_granges %>% 
  filter(gene_biotype == "ncRNA") %>%
  pull(gene_id) %>%
  unique ->
  ncrna_geneids

unenriched_norm_counts %>%
  filter(geneid %in% ncrna_geneids) %>%
  select(-geneid) %>%
  calculateCrossCorrelation(dds) ->
  ncrna_unenriched_cor

norm_counts_wmeans %>%
  filter(geneid %in% ncrna_geneids) %>%
  enrichedCorrelationToGoldStandard(dds) ->
  ncrna_enriched_cor

correlationRanges(ncrna_unenriched_cor, ncrna_enriched_cor) %>%
  kable
```

```{r}
plotCorrelationValues(ncrna_unenriched_cor, 
                      ncrna_enriched_cor, 
                      dds, fig_palette, enrich_levels)
ggsave(file.path(fig.dir, "Figure_4.tiff"), compression = "lzw", width=15, height=10.71, units = "cm")
```

```{r fig.width=8, fig.height=8}  
norm_counts_long_noreps %>%
  filter(geneid %in% ncrna_geneids) %>%
  makeEnrichmentScatterplot(fig_palette, enrich_levels)
ggsave(file.path(fig.dir, "Figure_S3.tiff"), compression = "lzw", width=20, height=20, units = "cm")
```

### Exploring RNaseH ncRNA outlier (CNAG_12993)
```{r}
norm_counts_long_noreps %>%
  filter(geneid %in% ncrna_geneids) %>%
  filter(enrichment_method=="RNase H") %>%
  arrange(desc(gold_mean)) %>%
  head(20) %>%
  kable
```

```{r}
norm_counts_wmeans %>%
  filter(geneid %in% ncrna_geneids) %>%
  filter(geneid != "CNAG_12993") %>%
  enrichedCorrelationToGoldStandard(dds) ->
  ncrna_enriched_drop_CNAG_12993_cor

unenriched_norm_counts %>%
  filter(geneid %in% ncrna_geneids) %>%
  filter(geneid != "CNAG_12993") %>%
  select(-geneid) %>%
  calculateCrossCorrelation(dds) ->
  ncrna_unenriched_drop_CNAG_12993_cor

plotCorrelationValues(ncrna_unenriched_drop_CNAG_12993_cor, 
                      ncrna_enriched_drop_CNAG_12993_cor, 
                      dds, fig_palette, enrich_levels)
ggsave(file.path(fig.dir, "Figure_S5.tiff"), compression = "lzw", width=15, height=10.71, units = "cm")
```

```{r}
correlationRanges(ncrna_unenriched_cor, ncrna_enriched_drop_CNAG_12993_cor) %>%
  kable
```

#### Does size explain CNAG_12993?

```{r}
import(gtf_with_mito_rrna.file, format="gtf") %>%
  as.data.frame %>%
  filter(gene_id %in% ncrna_geneids) %>%
  filter(type == "gene") %>%
  arrange(width) %>%
  head(20)
```

```{r}
import(gtf_with_mito_rrna.file, format="gtf") %>%
  as.data.frame %>%
  filter(gene_id %in% ncrna_geneids) %>%
  filter(type == "gene") %>%
  arrange(width) %>%
  ggplot(aes(x=width)) +   
   geom_histogram(aes(y=..density..)) + 
   geom_density(aes(y=..density..))

```

```{r}
import(gtf_with_mito_rrna.file, format="gtf") %>%
  as.data.frame %>%
  filter(gene_id %in% ncrna_geneids) %>%
  filter(type == "gene") ->
  gene_annotations

results(dds, contrast=c("enrichment_method","Unenriched","RNase H")) %>%
  as.data.frame %>%
  rownames_to_column("gene_id") %>%
  filter(gene_id %in% ncrna_geneids) %>%
  left_join(gene_annotations, by="gene_id") %>%
  ggplot(aes(width, log2FoldChange)) +
  geom_point()
```

# Pre-filtering to drop genes <50 reads in gold standard
```{r}
unenriched_low_counts = 50
unenriched_low_counts

keep <- rowSums(counts(dds[enrichment_method = 'Unenriched'])) >= unenriched_low_counts
dds_no_low_counts <- dds[keep,]
dds_no_low_counts
```
#DE analysis to ID genes with decreased expression in Poly(A) compared to Unenriched
```{r}
lfc_cutoff = log2(8)
sig_cutoff = 0.05

dea_polya_unenriched <- results(dds_no_low_counts, contrast = c("enrichment_method", "Poly(A)", "Unenriched"), lfcThreshold=lfc_cutoff, altHypothesis = "less")
dea_polya_unenriched

dea_polya_unenriched %>% 
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  arrange(padj, log2FoldChange) %>%
  filter(padj < sig_cutoff) %>%
  pull(gene_id) ->
  lost_polya_genes

norm_counts_long_noreps %>%
  mutate(polya_lost = geneid %in% lost_polya_genes) ->
  norm_counts_long_noreps_lost_polya

norm_counts_long_noreps_lost_polya %>%
  head %>%
  kable
```

#Scatterplot of genes with decreased expression in Poly(A) compared to Unenriched
```{r fig.width=8, fig.height=8}
norm_counts_long_noreps_lost_polya %>%
    ggplot(aes(x=gold_mean, y=norm_count, color=polya_lost)) +
    geom_point(alpha=0.3) +
    labs(x = "Average Normalized Count in Unenriched", y= "Normalized Count") +
    theme_bw() +
    theme(text=element_text(size=10)) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())  +
    facet_grid(enrichment_method ~ RNA_sample_num) +
    scale_colour_manual(values = c("FALSE"="grey60","TRUE"="red"), guide = FALSE) +
    geom_abline(size=0.5)

ggsave(file.path(fig.dir, "Figure_S6.tiff"), compression = "lzw", width=20, height=20, units = "cm")
```
```{r}
import(gtf_with_mito_rrna.file, format = "gtf") %>%
  as.data.frame %>%
  filter(gene_id %in% lost_polya_genes) %>%
  filter(type=="gene") %>%
  select(Gene=gene_id, Chromosome=seqnames, `Gene type`=gene_biotype) %>%
  arrange(`Gene type`, Chromosome) -> genes_lost_in_polya

genes_lost_in_polya %>%
  write.csv(file.path(fig.dir, "Table_S1.csv"), row.names = FALSE)

```

# SessionInfo
```{r}
sessionInfo()
```

