---
title: "Download Data"
output:
  html_document:
    toc: false
---

```{r}
source("run_config.R")

library(GEOquery)
library(dplyr)
library(rentrez)
library(readr)
library(stringr)
library(fs)
library(R.utils)
library(purrr)
library(tibble)
library(tools)
```

# Make Metadata Table with SRA and GEO Accesion Numbers
## Get SRA Metadata
```{r}
entrez_search(db="sra", term="PRJNA673133",retmax=1000) ->
  sra_search

# check that we got everything
stopifnot(sra_search[["count"]]==length(sra_search[["ids"]]))

entrez_fetch(db="sra", id=sra_search[["ids"]], rettype="runinfo") %>%
  read_csv %>%
  dplyr::select(Run, Experiment, SampleName) %>%
  filter(Run!="Run") ->
  sra_df
sra_df
```
## Get GEO Metadata
```{r}
getGEO("GSE160397",GSEMatrix=TRUE) ->
  gse
pData(gse[[1]]) %>%
  dplyr::select(geo_accession, title, `enrichment method:ch1`, `genotype:ch1`, `media:ch1`, `rna sample number:ch1`) %>%
  rename_with(str_remove, .cols = everything(), ":ch1") %>%
  rename_with(str_replace_all, .cols = everything(), " ", "_") %>%
  filter(rna_sample_number %in% c("2", "3", "4")) %>%
  mutate(sequencing_batch=ifelse(str_detect(title, "_2018_"),
                                "2019",
                                "2018")) %>%
  arrange(sequencing_batch, enrichment_method, rna_sample_number) ->
  geo_df
geo_df
```

## Join SRA and GEO Metadata
```{r}
right_join(sra_df, 
          geo_df, 
          by=c("SampleName"="geo_accession")) ->
  accessions_with_meta
accessions_with_meta
```





# Download FASTQs from SRA
## Deal with vdb-config



```{r}
fasterqDump = function(accession, outdir, tempdir=NULL, gzip=TRUE){
  fasterq_args =  c(accession, "--outdir", outdir)
  if (!is.null(tempdir)){
    # dir.create(tempdir, recursive = TRUE)
    fasterq_args =  c(fasterq_args, "--temp", tempdir)
  }
  system2(command="fasterq-dump",
          args=fasterq_args)
  
  file.path(outdir, accession) %>%
    path_ext_set("fastq") ->
    fastq_path
  if(gzip==TRUE){
    fastq_path=gzip(fastq_path)
  }
  return(fastq_path)
}
```


```{r eval=FALSE, include=FALSE}
fasterqDump(accession="SRR12933702",
            outdir=sra_dir, 
            tempdir="/workspace/tmp" )
```

```{r eval=FALSE, include=FALSE}
accessions_with_meta %>%
  pull(Run) %>%
  head(2) %>%
  map(function(x) fasterqDump(accession=x, outdir=sra_dir, tempdir="/workspace/tmp")) ->
  downloaded_fastqs
```


```{r}
accessions_with_meta %>%
  pull(Run) %>%
  map(function(x) fasterqDump(accession=x, outdir=sra_dir, tempdir="/workspace/tmp")) ->
  downloaded_fastqs
```


```{r}
downloaded_fastqs %>%
  unlist %>%
  md5sum %>%
  enframe %>%
  dplyr::rename(fullpath=name, md5sum=value) %>%
  mutate(filename=basename(fullpath)) %>%
  select(md5sum, filename) %>%
  write_delim(file.path(sra_dir, "md5_checksums.txt"), col_names = FALSE)
```



```{r eval=FALSE, include=FALSE}
accessions_with_meta %>%
  pull(Run) %>%
  head(2)
```

```{r eval=FALSE, include=FALSE}
fasterqDump(accession="SRR12933557",
            outdir=sra_dir, 
            tempdir="/workspace/tmp" )
```

# STOPPED HERE
```{r}
warning("Deal with vdb-config")
```
```{r}
stop("STOPPED HERE")
```

```{bash}
cat  ~/.ncbi/user-settings.mkfg 
```

```{r eval=FALSE, include=FALSE}
stop("Deal with vdb-config")
```

# Download Data
  - https://www.ncbi.nlm.nih.gov/home/tools/
  - https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE160397
  - https://www.biostars.org/p/111040/
  - https://bioinformaticsworkbook.org/dataAcquisition/fileTransfer/sra.html#gsc.tab=0
  - https://www.biostars.org/p/339182/
  - https://www.biostars.org/p/111040/#113204
  - https://www.biostars.org/p/160576/

## This Image
  - CRAN geomdb <https://cran.r-project.org/web/packages/geomedb/index.html>?
  - esearch
  - GEOquery <https://bioconductor.org/packages/release/bioc/html/GEOquery.html>
  - SRAdb <https://www.bioconductor.org/packages/release/bioc/html/SRAdb.html>


# SessionInfo
```{r}
sessionInfo()
```
