---
title: "Mito coverage plots"
output:
  html_document:
    toc: false
---

# Setup
## Load Libraries
```{r}
library(Rsamtools)
library(ggbio)
library(fs)
library(GenomicAlignments)
library(rtracklayer)
# library(tidyverse)
library(plyr)
library(stringr)
library(tibble)
library(dplyr)
library(readr)

source("run_config.R")
```

## Load Data
```{r}
# get paths from source run_star_config.sh?
h99_genome = readDNAStringSet(fa.file)

# Strip descriptions from chrom names
h99_genome %>%
  names %>%
  str_extract("\\w+") ->
  names(h99_genome)

# full_mito_name = str_subset(names(h99_genome), "Mt") # pull the full name of the mito sequence from fasta
# short_mito_name = str_extract(full_mito_name, "\\w+")
short_mito_name = "Mt"
small_mito_rrna_id = "15s_mito_rrna"
large_mito_rrna_id = "21s_mito_rrna"
```

## Define Genomic Ranges
```{r}
whole_mito_gr = GRanges(seqnames = "Mt", 
                        IRanges(1, 24919), strand = "*")

range_21s_region = GRanges(seqnames = short_mito_name, 
                              IRanges(6500, 9500), strand = "*")

range_15s_region = GRanges(seqnames = short_mito_name, 
                              IRanges(16950, 18500), strand = "*")
```




```{r}
accessions_file %>%
  read_csv %>%
  filter(sequencing_batch=="2018", 
         enrichment_method=="Unenriched") ->
  tot_2018_metadata

tot_2018_metadata %>%
  group_by(sample_id) %>%
  slice(1) %>%
  pull(Run) %>%
  paste0("_Aligned.sortedByCoord.out.bam") %>%
  file.path(total_samples_starout.dir,.) ->
  tot_L1_bam_files

tot_2018_metadata %>%
  pull(Run) %>%
  paste0("_Aligned.sortedByCoord.out.bam") %>%
  file.path(total_samples_starout.dir,.) ->
tot_bam_files

tot_bams = BamFileList(tot_bam_files, paste0(tot_bam_files, ".bai"))
tot_L1_bams = BamFileList(tot_L1_bam_files, paste0(tot_L1_bam_files, ".bai"))

tot_bam_files
```

# Whole Genome
```{r}
autoplot(tot_L1_bams, which = whole_mito_gr, geom = "area") +
  xlim(whole_mito_gr) +
  xlab("Position on Mitochondrial Chromosome") +
  theme_classic() ->
  whole_mito_plot

# ggsave(file.path(fig.dir, "whole_mito_rrna_coverage.png"), whole_mito_plot)
whole_mito_plot
```


# 21s rRNA region
```{r}
autoplot(tot_L1_bams, which = range_21s_region) +
  xlim(range_21s_region) ->
  mito_21S_plot

# ggsave(file.path(fig.dir, "mito_21S_rrna_coverage.png"), mito_21S_plot)
mito_21S_plot
```

## 21s rRNA Left End

```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(6705, 6740), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```

```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(6705, 6712), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```

```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(6705, 6710), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```

```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(6705, 6711), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```
left end looks like 6710


## 21s rRNA Right End
```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(9000, 9500), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```

```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(9300, 9329), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range) +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 20))
```
Right end looks like 9326

## Check Strand

```{r eval=FALSE, include=FALSE}
# not sure why this chunk isn't working now - it did previously! Maybe problem with BiocParallel?
region_21s_plus = GRanges(seqnames = short_mito_name, 
                              IRanges(6500, 9500), strand = "+")
region_21s_plus.overlaps = summarizeOverlaps(region_21s_plus, tot_L1_bams)
assay(region_21s_plus.overlaps)
```

```{r eval=FALSE, include=FALSE}
# not sure why this chunk isn't working now - it did previously!
region_21s_minus = GRanges(seqnames = short_mito_name, 
                              IRanges(6500, 9500), strand = "-")
region_21s_minus.overlaps = summarizeOverlaps(region_21s_minus, tot_L1_bams)
assay(region_21s_minus.overlaps)
```
Vast majoring of reads are + strand, so 21s rRNA gene must be - strand (since dUTP method was used)

## 21s rRNA annotation
```{r}
gene_21s = GRanges(seqnames = short_mito_name, 
                   IRanges(6710, 9326), strand = "-")
gene_21s
names(gene_21s) = "large_MTrRNA"

```






# 15s rRNA region
```{r}
autoplot(tot_L1_bams, which = range_15s_region) +
  xlim(range_15s_region) ->
  mito_15S_plot

# ggsave(file.path(fig.dir, "mito_15S_rrna_coverage.png"), mito_15S_plot)
mito_15S_plot
```

## 15s rRNA Left End

```{r}
range_15s_region = GRanges(seqnames = short_mito_name, 
                              IRanges(17000, 17175), strand = "*")

autoplot(tot_L1_bams, which = range_15s_region) +
  xlim(range_15s_region)
```

```{r}
range_15s_region = GRanges(seqnames = short_mito_name, 
                              IRanges(17000, 17100), strand = "*")

autoplot(tot_L1_bams, which = range_15s_region) +
  xlim(range_15s_region)
```

```{r}
range_15s_region = GRanges(seqnames = short_mito_name, 
                              IRanges(16940, 16980), strand = "*")

autoplot(tot_L1_bams, which = range_15s_region) +
  xlim(range_15s_region)  +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 20))
```

left end looks like 16948?


## 15s rRNA Right End
```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(18000, 18500), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```


```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(18290, 18315), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```

```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(18310, 18315), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```

```{r}
cur_range = GRanges(seqnames = short_mito_name, 
                              IRanges(18314, 18317), strand = "*")

autoplot(tot_L1_bams, which = cur_range) +
  xlim(cur_range)
```
Right end looks like 18316

## Check Strand
left end looks like 16948?
Right end looks like 18316

```{r eval=FALSE, include=FALSE}
# not sure why this chunk isn't working now - it did previously!
region_15s_plus = GRanges(seqnames = short_mito_name, 
                              IRanges(16948, 18316), strand = "+")
region_15s_plus.overlaps = summarizeOverlaps(region_15s_plus, tot_L1_bams)
assay(region_15s_plus.overlaps)
```

```{r eval=FALSE, include=FALSE}
# not sure why this chunk isn't working now - it did previously!
region_15s_minus = GRanges(seqnames = short_mito_name, 
                              IRanges(16948, 18316), strand = "-")
region_15s_minus.overlaps = summarizeOverlaps(region_15s_minus, tot_L1_bams)
assay(region_15s_minus.overlaps)
```
Vast majoring of reads are - strand, so 15s rRNA gene must be + strand

## 15s rRNA annotation
```{r}
gene_15s = GRanges(seqnames = short_mito_name, 
                   ranges=IRanges(16948, 18316), strand = "+")
names(gene_15s) = "small_MTrRNA"
```

# Extract and output 15s and 21s rRNA sequences


```{r}

seqnames(gene_15s)
# seq_15s = getSeq(h99_genome, mito_name)
seq_15s = getSeq(h99_genome, gene_15s)
names(seq_15s) = "small_mito_rRNA__CNA3"
seq_21s = getSeq(h99_genome, gene_21s)
names(seq_21s) = "large_mito_rRNA__CNA3"

writeXStringSet(seq_15s, file.path(seq.dir, "small_mito_rRNA__CNA3.fasta"), format="fasta")
writeXStringSet(seq_21s, file.path(seq.dir, "large_mito_rRNA__CNA3.fasta"), format="fasta")
```

Fred's h99 mito anotation: https://www.ncbi.nlm.nih.gov/nucleotide/AY101381.1

# Export rRNA gene coordinates as GTF
https://support.bioconductor.org/p/70054/

```{r eval=FALSE, include=FALSE}
# ? export.gff
mito_rrna_gtf.file = file.path(seq.dir, "mito_rrna.gtf")

export(GRangesList(gene_15s, gene_21s), 
       con = mito_rrna_gtf.file,
       format="gff2")

read_lines(mito_rrna_gtf.file)
```

```{r eval=FALSE, include=FALSE}
h99_gtf = read_tsv(gtf.file,
                   col_names = c("seqname", "source", "feature", "start", "end", "score", 
                       "strand", "frame", "attributes"),
                   cols(seqname = col_character(),
                        source = col_character(),
                        feature = col_character(),
                        start = col_double(),
                        end = col_double(),
                        score = col_character(),
                        strand = col_character(),
                        frame = col_character(),
                        attributes = col_character()
                        ),
                   comment = "#"
         )

mito_rrna_gtf = read_tsv(mito_rrna_gtf.file,
                   col_names = c("seqname", "source", "feature", "start", "end", "score", 
                       "strand", "frame", "attributes"),
                   cols(seqname = col_character(),
                        source = col_character(),
                        feature = col_character(),
                        start = col_double(),
                        end = col_double(),
                        score = col_character(),
                        strand = col_character(),
                        frame = col_character(),
                        attributes = col_character()
                        ),
                   comment = "#"
         )
```

```{r eval=FALSE, include=FALSE}
h99_gtf %>%
  filter(str_detect(attributes, "CNAG_10500")) 
```

```{r eval=FALSE, include=FALSE}
h99_gtf %>%
  filter(str_detect(attributes, 'gene_biotype "rRNA"')) %>%
  filter(feature=="exon")
```

```{r eval=FALSE, include=FALSE}
h99_gtf %>%
  filter(str_detect(attributes, 'gene_biotype "rRNA"')) %>%
  filter(feature=="exon") %>%
  select(attributes) ->
  x
x[[1]]
```

```{r}
file_copy(gtf.file, gtf_with_mito_rrna.file, overwrite=TRUE)
```

```{r}
make_df_from_gr = function(gr_vec){
  attr_template = 'gene_id "XGENEIDX"; transcript_id "XGENEIDX-1"; exon_number "1"; gene_name "XGENEIDX"; gene_source "JAG"; gene_biotype "rRNA"; transcript_name "XGENEIDX"; transcript_source "JAG"; transcript_biotype "rRNA"; exon_id "XGENEIDX-1-1";'
  attr = str_replace_all(attr_template, "XGENEIDX", names(gr_vec))
  
  names(gr_vec)
  print(attr_template)
  print(attr)
  gr_vec %>%
    as_tibble %>%
    add_column(attributes=attr, source="JAG", feature="exon", score=".", frame=".") %>%
    dplyr::select(seqname=seqnames, source, feature, start, end, score, 
                       strand, frame, attributes)
    
}

mito_rrdna_gtf.gf = ldply(GRangesList(gene_15s, gene_21s), make_df_from_gr)
# write_tsv(mito_rrdna_gtf.gf, gtf_with_mito_rrna.file, append = TRUE, quote_escape=FALSE)
write.table(mito_rrdna_gtf.gf, 
            gtf_with_mito_rrna.file, 
            append = TRUE, 
            quote = FALSE, sep = "\t",
            row.names = FALSE,
            col.names = FALSE)
file_copy(gtf_with_mito_rrna.file, file_S2_path, overwrite = TRUE)
```

```{bash}
set -u
grep -i rRNA $GTF
```

```{bash}
grep CNAG_10500 $TOTAL_SAMPLES_STAROUT/*_ReadsPerGene.out.tab
```

# Figures for Manuscript
## Whole Genome
```{r fig.width=8, fig.height=4}
autoplot(tot_L1_bams[[1]], which = whole_mito_gr, geom="area") +
  xlim(whole_mito_gr) +
  xlab("Position on Mitochondrial Chromosome") +
  ylab("Depth of Coverage") +
  theme_classic(base_size = 10) ->
  whole_mito_plot

whole_mito_plot@ggplot
ggsave(file.path(fig.dir, "Figure_S1.tiff"), plot=whole_mito_plot@ggplot, compression = "lzw")
```


# Notes
## BioPlots
### Mirror Plots
 - [Question: ggbio: mirrored coverage plots for +/- strands](https://support.bioconductor.org/p/56456/)

## Gviz
  - [Plot Sequencing Depth with Gviz](https://blog.liang2.tw/posts/2016/01/plot-seq-depth-gviz/)
  - [The Gviz User Guide](https://bioconductor.org/packages/release/bioc/vignettes/Gviz/inst/doc/Gviz.pdf)

# Play

```{r eval=FALSE, include=FALSE}
cur_range = GRanges(seqnames = short_mito_name,
                              IRanges(6500, 9500), strand = "*")

curbam = tot_bams[[1]]
autoplot(curbam, which = cur_range, color="red") +
  xlim(cur_range) +
  ggtitle(basename(curbam$path))

curbam = tot_bams[[5]]
autoplot(curbam, which = cur_range, color="red") +
  xlim(cur_range) +
  ggtitle(basename(curbam$path))
```


# SessionInfo
```{r}
sessionInfo()
```
