---
title: "Run STAR on TOTAL samples"
output:
  html_document:
    toc: false
---

```{r}
source("run_config.R")

library(readr)
library(dplyr)
```

```{r}
accessions_file %>%
  read_csv %>%
  filter(sequencing_batch=="2018") %>%
  pull(Run) %>%
  paste0(".fastq.gz") %>%
  file.path(sra_dir,.) ->
  all_2018_fastq_paths

all_2018_fastq_paths

all_2018_fastq_paths %>%
  paste(collapse = " ") %>%
  Sys.setenv(ALL_2018_FASTQS=.)
```

```{bash}
set -u
mkdir -p ${ENRICH_COMPARE_STAROUT}
trim_and_star_func() {
    FASTQ=$1
    SAMPLE="$(basename ${FASTQ} .fastq.gz)" # directory and strip .fastq.gz file extension
    FASTQ_TRIMMED="$TRIMMED/${SAMPLE}_trimmed.fq.gz"
    echo $SAMPLE
    echo $FASTQ

    # make a pipe for trimmed fastq
    # CUR_PIPE=`mktemp --dry-run`_${SAMPLE}_pipe.fq
    # mkfifo $CUR_PIPE

    # Run fastq-mcf
    fastq-mcf \
        $ADAPTERS \
        $FASTQ \
        -o $FASTQ_TRIMMED \
        -q 20 -x 0.5
        
    STAR \
    --runMode alignReads \
    --runThreadN $THREADS \
    --genomeDir $GENOME_WITH_MITO_RRNA_DIR \
    --outSAMtype BAM SortedByCoordinate \
    --quantMode GeneCounts \
    --genomeLoad NoSharedMemory \
    --twopassMode None \
    --limitBAMsortRAM 1280000000 \
    --outFileNamePrefix ${ENRICH_COMPARE_STAROUT}/${SAMPLE}_ \
    --outFilterScoreMinOverLread 0 --outFilterMatchNminOverLread 0 \
    --outFilterMatchNmin 0 --outFilterMismatchNmax 2 \
    --alignIntronMax 1000 --alignMatesGapMax 1000 \
    --readFilesCommand zcat \
    --readFilesIn $FASTQ_TRIMMED        
        
    # rm -f $CUR_PIPE
    samtools index ${ENRICH_COMPARE_STAROUT}/${SAMPLE}_Aligned.sortedByCoord.out.bam
}
export -f trim_and_star_func

parallel --jobs $MAX_JOBS trim_and_star_func ::: $ALL_2018_FASTQS
# This needs to be here to knit. Knitting fails if the line above is the last line in the chunk!
```

### Check that custom GTF (with MTrRNA) was used
```{bash}
grep MTrRNA $ENRICH_COMPARE_STAROUT/*_ReadsPerGene.out.tab
```

# SessionInfo
```{r}
sessionInfo()
```
